---
title: "Fitting for bulk DNA-seq data in HGSOC"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Fitting for bulk DNA-seq data in HGSOC}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  warning=FALSE, message=FALSE,
  comment = "#>"
)
```

## Background
This vignette illustrates how to use the program to fit driver, selection and CN event rates for a model.
First we install the simulator...

```{r setup}
devtools::install_github("dinhngockhanh/CancerSimulator")

library(ggplot2)
library(signals)
library(CancerSimulator)

devtools::load_all()
devtools::document()
```







## Set up the model
For this example, we simulate the mutation-driven evolution of HGSOC.

```{r}
MODEL <- 'HGSOC-BULK'
SIMULATOR_VARIABLES_for_simulation(MODEL)
```

We build an ABC to fit the driver mutation rate and CN event rates:

```{r}
cat('Poisson rate of new driver mutations                         = ',rate_driver,'\n')
cat('Probability of WGD                                           = ',prob_CN_whole_genome_duplication,'\n')
cat('Probability of chromosome missegragation                     = ',prob_CN_missegregation,'\n')
cat('Probability of chromosome-arm missegragation                 = ',prob_CN_chrom_arm_missegregation,'\n')
cat('Probability of focal amplification                           = ',prob_CN_focal_amplification,'\n')
cat('Probability of focal deletion                                = ',prob_CN_focal_deletion,'\n')
cat('Probability of interstitial CN-LOH                           = ',prob_CN_cnloh_interstitial,'\n')
cat('Probability of terminal CN-LOH                               = ',prob_CN_cnloh_terminal,'\n')
cat('Geometric parameter for bin length of a focal amplification  = ',prob_CN_focal_amplification_length,'\n')
cat('Geometric parameter for bin length of a focal deletion       = ',prob_CN_focal_deletion_length,'\n')
cat('Geometric parameter for bin length of an interstitial CN-LOH = ',prob_CN_cnloh_interstitial_length,'\n')
cat('Geometric parameter for bin length of a terminal CN-LOH      = ',prob_CN_cnloh_terminal_length,'\n')
```

as well as the selection rates for each mutational driver found in the data:

```{r}
driver_library
```










## Statistics for a parameter set
Assume that we have a parameter set for the rates:
```{r}
vec_selection_rates                                   <- rep(0.05,nrow(driver_library))

rate_driver                                           <- 1e-16

prob_CN_whole_genome_duplication                      <- 1e-6
prob_CN_missegregation                                <- 1e-6
prob_CN_chrom_arm_missegregation                      <- 1e-6
prob_CN_focal_amplification                           <- 1e-6
prob_CN_focal_deletion                                <- 1e-6
prob_CN_cnloh_interstitial                            <- 1e-6
prob_CN_cnloh_terminal                                <- 1e-6

prob_CN_focal_amplification_length                    <- 0.1
prob_CN_focal_deletion_length                         <- 0.1
prob_CN_cnloh_interstitial_length                     <- 0.1
prob_CN_cnloh_terminal_length                         <- 0.1

parameter_set                                         <- list()
parameter_set$vec_selection_rates                     <- vec_selection_rates
parameter_set$rate_driver                             <- rate_driver
parameter_set$prob_CN_whole_genome_duplication        <- prob_CN_whole_genome_duplication
parameter_set$prob_CN_missegregation                  <- prob_CN_missegregation
parameter_set$prob_CN_chrom_arm_missegregation        <- prob_CN_chrom_arm_missegregation
parameter_set$prob_CN_focal_amplification             <- prob_CN_focal_amplification
parameter_set$prob_CN_focal_deletion                  <- prob_CN_focal_deletion
parameter_set$prob_CN_cnloh_interstitial              <- prob_CN_cnloh_interstitial
parameter_set$prob_CN_cnloh_terminal                  <- prob_CN_cnloh_terminal
parameter_set$prob_CN_focal_amplification_length      <- prob_CN_focal_amplification_length
parameter_set$prob_CN_focal_deletion_length           <- prob_CN_focal_deletion_length
parameter_set$prob_CN_cnloh_interstitial_length       <- prob_CN_cnloh_interstitial_length
parameter_set$prob_CN_cnloh_terminal_length           <- prob_CN_cnloh_terminal_length
```

The following function sets up the model variables according to the input parameter set:

```{r}
SIMULATOR_VARIABLES_for_fitting(MODEL,parameter_set)
```

Specifically, it translates the input selection strengths for mutational drivers into selection rates for the WT and MUT alleles:

```{r}
driver_library
```



```{r}
ABC_one_simulation(MODEL,parameter_set)
```
