<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>An introduction to CINner</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">An introduction to CINner</h1>



<p>We first load <code>CINner</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(CINner)</span></code></pre></div>
<div id="set-up-model-variables" class="section level1">
<h1>Set up model variables</h1>
<p>CINner requires many input parameters about the assumed
characteristics of cancer growth. One easy method to load the parameters
is by creating <code>model_variables</code>, which is a list containing
relevant variables in a specific order, then using function
<code>SAVE_model_variables</code> to save them as csv files that can be
read in the next sections when the simulator is called. Here we detail
how to create and the parameter list.</p>
<div id="general-variables-for-every-cell" class="section level2">
<h2>General variables for every cell</h2>
<p>We first specify the cell lifespan (always in days). This is the mean
of the exponentially distributed lifetime of every cell during cancer
evolution, regardless of their driver mutation or CN profiles.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>cell_lifespan <span class="ot">&lt;-</span> <span class="dv">30</span></span></code></pre></div>
<p>Next, we specify the time points at the beginning and the end of the
simulation. The unit can be <code>day</code>, <code>week</code> (= 7
days), <code>month</code> (= 30 days) or <code>year</code> (= 365 days).
The unit should be the same for both time points.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>T_0 <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="dv">0</span>, <span class="st">&quot;year&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>T_end <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="dv">80</span>, <span class="st">&quot;year&quot;</span>)</span></code></pre></div>
<p>Next variables dictate the time points that each sample is taken
(same unit as the beginning and end time points), how many cells are in
each sample, and the unique ID for each sample.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>Table_sample <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Sample_ID =</span> <span class="st">&quot;SA01&quot;</span>, <span class="at">Cell_count =</span> <span class="dv">300</span>, <span class="at">Age_sample =</span> <span class="dv">80</span>)</span></code></pre></div>
<p>The simulator’s efficiency is based partially on its use of the
tau-leaping algorithm, which simulates the evolution in fixed steps
instead of the exact Gillespie algorithm. As such, it requires a
suggested time step. If not specified, the time step is automatically
set to be half the cell lifespan. For statistical accuracy, it should
not be higher. However, setting significantly low numbers for time step
will lead to increased simulation run time.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>T_tau_step <span class="ot">&lt;-</span> cell_lifespan <span class="sc">/</span> <span class="dv">2</span></span></code></pre></div>
<p>To speed up the simulations and in accordance with the current state
of single-cell DNA sequencing, the genome is divided into bins of equal
lengths. Here we specify the length of each bin, understood in number of
base pairs.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>CN_bin_length <span class="ot">&lt;-</span> <span class="dv">500000</span></span></code></pre></div>
<p>Three selection models are included in the framework: -
<code>chrom-arm-selection</code> models the selection for gains and
losses of chromosome arms, depending on their genetic balances. -
<code>driver-gene-selection</code> models the selection for gains and
losses and mutations of specific driver genes. -
<code>chrom-arm-and-driver-gene-selection</code> combines both models
into a hybrid model, where cancer is driven both by large-scale genetic
imbalance and local events targeting specific genes.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>selection_model <span class="ot">&lt;-</span> <span class="st">&quot;chrom-arm-and-driver-gene-selection&quot;</span></span></code></pre></div>
<p>The next variables are the probabilities that each class of CNA
events occurs in a cell division. There are three large-scale CNA
mechanisms: - Whole Genome Duplication results in one daughter cell with
double the genomic material of the mother cell. - Missegregation targets
a uniformly chosen chromosome strand, which is then gained in one
daughter cell and lost in the other. - The same occurs for
chromosome-arm missegregation but only for one of the chromosome’s arms
with equal probabilities.</p>
<p>The local events target a randomly chosen region in a uniformly
chosen chromosome arm in one of the daughter cells. All of these events
require additional parameters for the region length distributions, to be
discussed next. There are four local CNA mechanisms: - Focal
amplification and deletion duplicate or delete the region, respectively.
- Interstitial and terminal CNLOH replace all alleles within the region
across chromosome strands with an allele chosen at random. Interstitial
CNLOH can occur anywhere in the arm, but terminal CNLOH has to occur at
the chromosome’s end.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>prob_CN_whole_genome_duplication <span class="ot">&lt;-</span> <span class="fl">0.5e-4</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>prob_CN_missegregation <span class="ot">&lt;-</span> <span class="fl">2e-4</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>prob_CN_chrom_arm_missegregation <span class="ot">&lt;-</span> <span class="fl">2e-4</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>prob_CN_focal_amplification <span class="ot">&lt;-</span> <span class="dv">0</span>.e<span class="dv">-4</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>prob_CN_focal_deletion <span class="ot">&lt;-</span> <span class="fl">2e-4</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>prob_CN_cnloh_interstitial <span class="ot">&lt;-</span> <span class="fl">0e-4</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>prob_CN_cnloh_terminal <span class="ot">&lt;-</span> <span class="fl">0e-4</span></span></code></pre></div>
<p>In specific applications, we might also consider neutral versions of
the aforementioned CNA mechanisms. These events are assumed to change
the CN profiles of the daughter cells without affecting their fitness.
The rules dictating how the daughter cells’ CN profiles change as a
result of neutral CNAs are similar to the selective CNA mechanisms
above, including the additional parameters specifying the region length
distributions of the local CNA events.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>prob_neutral_CN_whole_genome_duplication <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>prob_neutral_CN_missegregation <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>prob_neutral_CN_chrom_arm_missegregation <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>prob_neutral_CN_focal_amplification <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>prob_neutral_CN_focal_deletion <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>prob_neutral_CN_cnloh_interstitial <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>prob_neutral_CN_cnloh_terminal <span class="ot">&lt;-</span> <span class="dv">0</span></span></code></pre></div>
<p>We now specify the parameters for the region length distributions of
the local CNA mechanisms. Let <code>X</code> be a given local CNA
mechanism (<code>X</code> = <code>CN_focal_amplification_length</code>,
<code>CN_focal_deletion_length</code>,
<code>CN_cnloh_interstitial_length</code>,
<code>CN_cnloh_terminal_length</code>). We choose between two
distributions: - <code>model_X</code> = <code>geom</code>: the region
length is geometrically distributed, with the rate given by another
parameter <code>prob_X</code> (unit is bin counts). If
<code>model_X</code> is not specified, then this is the default. -
<code>model_X</code> = <code>beta</code>: the region length equals a
beta distributed ratio of the chosen chromosome arm. This requires two
parameters for the beta distribution, given by
<code>prob_X_shape_1</code> and <code>prob_X_shape_2</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>model_CN_focal_amplification_length <span class="ot">&lt;-</span> <span class="st">&quot;beta&quot;</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>prob_CN_focal_amplification_length_shape_1 <span class="ot">&lt;-</span> <span class="fl">0.758304780825031</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>prob_CN_focal_amplification_length_shape_2 <span class="ot">&lt;-</span> <span class="fl">5.33873409782625</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>model_CN_focal_deletion_length <span class="ot">&lt;-</span> <span class="st">&quot;beta&quot;</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>prob_CN_focal_deletion_length_shape_1 <span class="ot">&lt;-</span> <span class="fl">0.814054548726361</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>prob_CN_focal_deletion_length_shape_2 <span class="ot">&lt;-</span> <span class="fl">6.16614890284825</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>prob_CN_cnloh_interstitial_length <span class="ot">&lt;-</span> <span class="fl">0.005</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>prob_CN_cnloh_terminal_length <span class="ot">&lt;-</span> <span class="fl">0.005</span></span></code></pre></div>
<p>The rate of driver mutations is per base pair per division.
Therefore, cells with more genomic materials, for instance after Whole
Genome Duplication, would mutate at increased rates. If a driver
mutation occurs, the gene to acquire the new mutation is chosen from an
input list of driver genes, excluding the genes already mutated in the
cell. A cell’s fitness increases with the number of driver mutations it
has acquired.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>rate_driver <span class="ot">&lt;-</span> <span class="dv">0</span></span></code></pre></div>
<p>The rate of passenger mutations is likewise per base pair per
division. However, passenger mutations do not change a cell’s
fitness.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>rate_passenger <span class="ot">&lt;-</span> <span class="fl">1e-11</span></span></code></pre></div>
<p>A cell increases its fitness by acquiring CNAs and driver mutations.
However, the cell is above all subject to viability conditions. If it
violates any of these conditions, the cell is deemed unviable and its
fitness is set to 0, therefore will die: - <code>bound_driver</code> is
the upper bound on how many driver mutations the cell has acquired. -
<code>bound_average_ploidy</code> is the upper bound on the ploidy
(i.e. average CN among all bins) of the cell. -
<code>bound_maximum_CN</code> is the upper bound on the CN in any bin in
the cell. - <code>bound_homozygosity</code> is the upper bound on the
bin count with homozygosity (i.e. bins with CN=0).</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>bound_driver <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>bound_average_ploidy <span class="ot">&lt;-</span> <span class="fl">4.5</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>bound_maximum_CN <span class="ot">&lt;-</span> <span class="dv">8</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>bound_homozygosity <span class="ot">&lt;-</span> <span class="dv">500</span></span></code></pre></div>
<p>The algorithm is designed to follow any dynamics (within the limit
imposed by how fast the cells can divide or die, defined by their
lifespan). The expected population dynamics is input as two vectors:
<code>vec_cell_count</code> lists the requested total cell counts at the
corresponding time points listed in <code>vec_time</code>, and both
vectors are combined in variable <code>table_population_dynamics</code>.
The expected population size in the simulations at any time point is
interpolated from the two nearest time points in this variable.</p>
<p>Here, we request a logistic expansion from 20 cells at the beginning
time point to 10000 cells at the final time point.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>vec_time <span class="ot">&lt;-</span> T_0[[<span class="dv">1</span>]]<span class="sc">:</span>T_end[[<span class="dv">1</span>]]</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>vec_cell_count <span class="ot">&lt;-</span> <span class="dv">10000</span> <span class="sc">/</span> (<span class="dv">1</span> <span class="sc">+</span> <span class="fu">exp</span>(<span class="sc">-</span><span class="fl">0.3</span> <span class="sc">*</span> (vec_time <span class="sc">-</span> <span class="dv">20</span>)))</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>table_population_dynamics <span class="ot">&lt;-</span> <span class="fu">cbind</span>(vec_time, vec_cell_count)</span></code></pre></div>
<p>If noisy readcounts based on the cells’ true CN profiles are
requested, we need to also set the parameters for the sequencing noise
and bias model:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>gc <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="at">file =</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;gc_map_500kb.csv&quot;</span>, <span class="at">package =</span> <span class="st">&quot;CINner&quot;</span>))</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>gc_slope <span class="ot">&lt;-</span> <span class="fl">1.2</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>gc_int <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>sigma1 <span class="ot">&lt;-</span> <span class="fl">0.02642392</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>num_reads <span class="ot">&lt;-</span> <span class="dv">3906632</span></span></code></pre></div>
<p>Finally, using the function <code>BUILD_general_variables</code>, we
can build <code>model_variables</code>, an R list containing all
variables described thus far:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>model_variables <span class="ot">&lt;-</span> <span class="fu">BUILD_general_variables</span>(</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>    <span class="at">cell_lifespan =</span> cell_lifespan,</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>    <span class="at">T_0 =</span> T_0, <span class="at">T_end =</span> T_end, <span class="at">T_tau_step =</span> T_tau_step,</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>    <span class="at">Table_sample =</span> Table_sample,</span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>    <span class="at">CN_bin_length =</span> CN_bin_length,</span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a>    <span class="at">prob_CN_whole_genome_duplication =</span> prob_CN_whole_genome_duplication,</span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>    <span class="at">prob_CN_missegregation =</span> prob_CN_missegregation,</span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a>    <span class="at">prob_CN_chrom_arm_missegregation =</span> prob_CN_chrom_arm_missegregation,</span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a>    <span class="at">prob_CN_focal_amplification =</span> prob_CN_focal_amplification,</span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a>    <span class="at">prob_CN_focal_deletion =</span> prob_CN_focal_deletion,</span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a>    <span class="at">prob_CN_cnloh_interstitial =</span> prob_CN_cnloh_interstitial,</span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a>    <span class="at">prob_CN_cnloh_terminal =</span> prob_CN_cnloh_terminal,</span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a>    <span class="at">prob_neutral_CN_whole_genome_duplication =</span> prob_neutral_CN_whole_genome_duplication,</span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a>    <span class="at">prob_neutral_CN_missegregation =</span> prob_neutral_CN_missegregation,</span>
<span id="cb16-15"><a href="#cb16-15" tabindex="-1"></a>    <span class="at">prob_neutral_CN_chrom_arm_missegregation =</span> prob_neutral_CN_chrom_arm_missegregation,</span>
<span id="cb16-16"><a href="#cb16-16" tabindex="-1"></a>    <span class="at">prob_neutral_CN_focal_amplification =</span> prob_neutral_CN_focal_amplification,</span>
<span id="cb16-17"><a href="#cb16-17" tabindex="-1"></a>    <span class="at">prob_neutral_CN_focal_deletion =</span> prob_neutral_CN_focal_deletion,</span>
<span id="cb16-18"><a href="#cb16-18" tabindex="-1"></a>    <span class="at">prob_neutral_CN_cnloh_interstitial =</span> prob_neutral_CN_cnloh_interstitial,</span>
<span id="cb16-19"><a href="#cb16-19" tabindex="-1"></a>    <span class="at">prob_neutral_CN_cnloh_terminal =</span> prob_neutral_CN_cnloh_terminal,</span>
<span id="cb16-20"><a href="#cb16-20" tabindex="-1"></a>    <span class="at">model_CN_focal_amplification_length =</span> model_CN_focal_amplification_length,</span>
<span id="cb16-21"><a href="#cb16-21" tabindex="-1"></a>    <span class="at">model_CN_focal_deletion_length =</span> model_CN_focal_deletion_length,</span>
<span id="cb16-22"><a href="#cb16-22" tabindex="-1"></a>    <span class="at">prob_CN_focal_amplification_length_shape_1 =</span> prob_CN_focal_amplification_length_shape_1,</span>
<span id="cb16-23"><a href="#cb16-23" tabindex="-1"></a>    <span class="at">prob_CN_focal_amplification_length_shape_2 =</span> prob_CN_focal_amplification_length_shape_2,</span>
<span id="cb16-24"><a href="#cb16-24" tabindex="-1"></a>    <span class="at">prob_CN_focal_deletion_length_shape_1 =</span> prob_CN_focal_deletion_length_shape_1,</span>
<span id="cb16-25"><a href="#cb16-25" tabindex="-1"></a>    <span class="at">prob_CN_focal_deletion_length_shape_2 =</span> prob_CN_focal_deletion_length_shape_2,</span>
<span id="cb16-26"><a href="#cb16-26" tabindex="-1"></a>    <span class="at">prob_CN_cnloh_interstitial_length =</span> prob_CN_cnloh_interstitial_length,</span>
<span id="cb16-27"><a href="#cb16-27" tabindex="-1"></a>    <span class="at">prob_CN_cnloh_terminal_length =</span> prob_CN_cnloh_terminal_length,</span>
<span id="cb16-28"><a href="#cb16-28" tabindex="-1"></a>    <span class="at">rate_driver =</span> rate_driver,</span>
<span id="cb16-29"><a href="#cb16-29" tabindex="-1"></a>    <span class="at">rate_passenger =</span> rate_passenger,</span>
<span id="cb16-30"><a href="#cb16-30" tabindex="-1"></a>    <span class="at">selection_model =</span> selection_model,</span>
<span id="cb16-31"><a href="#cb16-31" tabindex="-1"></a>    <span class="at">bound_driver =</span> bound_driver,</span>
<span id="cb16-32"><a href="#cb16-32" tabindex="-1"></a>    <span class="at">bound_average_ploidy =</span> bound_average_ploidy,</span>
<span id="cb16-33"><a href="#cb16-33" tabindex="-1"></a>    <span class="at">bound_homozygosity =</span> bound_homozygosity,</span>
<span id="cb16-34"><a href="#cb16-34" tabindex="-1"></a>    <span class="at">table_population_dynamics =</span> table_population_dynamics,</span>
<span id="cb16-35"><a href="#cb16-35" tabindex="-1"></a>    <span class="at">gc =</span> gc,</span>
<span id="cb16-36"><a href="#cb16-36" tabindex="-1"></a>    <span class="at">gc_slope =</span> gc_slope,</span>
<span id="cb16-37"><a href="#cb16-37" tabindex="-1"></a>    <span class="at">gc_int =</span> gc_int,</span>
<span id="cb16-38"><a href="#cb16-38" tabindex="-1"></a>    <span class="at">sigma1 =</span> sigma1,</span>
<span id="cb16-39"><a href="#cb16-39" tabindex="-1"></a>    <span class="at">num_reads =</span> num_reads</span>
<span id="cb16-40"><a href="#cb16-40" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="selection-variables" class="section level2">
<h2>Selection variables</h2>
<p>Depending on which selection model is specified in the variable
<code>selection_model</code> to drive cancer evolution, we need to
define additional variables.</p>
<div id="chromosome-arm-selection-model" class="section level3">
<h3>Chromosome arm selection model</h3>
<p>If <code>selection_model</code> = <code>chrom-arm-selection</code> or
<code>chrom-arm-and-driver-gene-selection</code>, then cancer is driven
solely or partially by gains and losses of chromosome arms, based on
their gene imbalances.</p>
<p>Five vectors in combination establish the chromosome arm selection
rates. Given values at the same location in the vectors,
<code>arm_s</code> is the selection rate for the arm with ID
<code>arm_id</code>, which runs from starting bin <code>arm_start</code>
to ending bin <code>arm_end</code> on chromosome
<code>arm_chromosome</code>. The selection rate ranges from 0 to
infinity. Arms with more Tumor Suppressor Genes than Oncogenes have
selection rate less than 1, for which losses are advantageous and gains
are deleterious. Arms with more Oncogenes than Tumor Suppressor Genes
have selection rate higher than 1, for which gains are advantageous and
losses are deleterious. The selective increases or decreases of the
gains and losses of an arm depend on how far its selection rate is from
1.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>arm_id <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">paste</span>(model_variables<span class="sc">$</span>cn_info<span class="sc">$</span>Chromosome, <span class="st">&quot;p&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>), <span class="fu">paste</span>(model_variables<span class="sc">$</span>cn_info<span class="sc">$</span>Chromosome, <span class="st">&quot;q&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>arm_chromosome <span class="ot">&lt;-</span> <span class="fu">rep</span>(model_variables<span class="sc">$</span>cn_info<span class="sc">$</span>Chromosome, <span class="dv">2</span>)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>arm_start <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>, <span class="fu">length</span>(model_variables<span class="sc">$</span>cn_info<span class="sc">$</span>Chromosome)), model_variables<span class="sc">$</span>cn_info<span class="sc">$</span>Centromere_location <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>arm_end <span class="ot">&lt;-</span> <span class="fu">c</span>(model_variables<span class="sc">$</span>cn_info<span class="sc">$</span>Centromere_location, model_variables<span class="sc">$</span>cn_info<span class="sc">$</span>Bin_count)</span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>arm_s <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="fu">length</span>(arm_id))</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;1p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.05</span></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;1q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.065</span></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;2p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.06</span></span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;2q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.05</span></span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;3p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.02</span></span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;3q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.095</span></span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;4p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.013</span></span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;4q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.016</span></span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;5p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.18</span></span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;5q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.1</span></span>
<span id="cb17-16"><a href="#cb17-16" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;6p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.2</span></span>
<span id="cb17-17"><a href="#cb17-17" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;6q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.17</span></span>
<span id="cb17-18"><a href="#cb17-18" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;7p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.02</span></span>
<span id="cb17-19"><a href="#cb17-19" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;7q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.1</span></span>
<span id="cb17-20"><a href="#cb17-20" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;8p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.14</span></span>
<span id="cb17-21"><a href="#cb17-21" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;8q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.2</span></span>
<span id="cb17-22"><a href="#cb17-22" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;9p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.022</span></span>
<span id="cb17-23"><a href="#cb17-23" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;9q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.022</span></span>
<span id="cb17-24"><a href="#cb17-24" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;10p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.03</span></span>
<span id="cb17-25"><a href="#cb17-25" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;10q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.035</span></span>
<span id="cb17-26"><a href="#cb17-26" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;11p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.15</span></span>
<span id="cb17-27"><a href="#cb17-27" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;11q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.13</span></span>
<span id="cb17-28"><a href="#cb17-28" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;12p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.08</span></span>
<span id="cb17-29"><a href="#cb17-29" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;12q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.04</span></span>
<span id="cb17-30"><a href="#cb17-30" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;13p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.03</span></span>
<span id="cb17-31"><a href="#cb17-31" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;13q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.022</span></span>
<span id="cb17-32"><a href="#cb17-32" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;14p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.022</span></span>
<span id="cb17-33"><a href="#cb17-33" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;14q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.016</span></span>
<span id="cb17-34"><a href="#cb17-34" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;15p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.02</span></span>
<span id="cb17-35"><a href="#cb17-35" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;15q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.01</span></span>
<span id="cb17-36"><a href="#cb17-36" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;16p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.01</span></span>
<span id="cb17-37"><a href="#cb17-37" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;16q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.08</span></span>
<span id="cb17-38"><a href="#cb17-38" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;17p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.07</span></span>
<span id="cb17-39"><a href="#cb17-39" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;17q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.01</span></span>
<span id="cb17-40"><a href="#cb17-40" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;18p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.04</span></span>
<span id="cb17-41"><a href="#cb17-41" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;18q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.03</span></span>
<span id="cb17-42"><a href="#cb17-42" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;19p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.06</span></span>
<span id="cb17-43"><a href="#cb17-43" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;19q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.06</span></span>
<span id="cb17-44"><a href="#cb17-44" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;20p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.05</span></span>
<span id="cb17-45"><a href="#cb17-45" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;20q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.06</span></span>
<span id="cb17-46"><a href="#cb17-46" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;21p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.22</span></span>
<span id="cb17-47"><a href="#cb17-47" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;21q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fl">0.08</span></span>
<span id="cb17-48"><a href="#cb17-48" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;22p&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.02</span></span>
<span id="cb17-49"><a href="#cb17-49" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;22q&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.02</span></span>
<span id="cb17-50"><a href="#cb17-50" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;Xp&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.04</span></span>
<span id="cb17-51"><a href="#cb17-51" tabindex="-1"></a>arm_s[<span class="fu">which</span>(arm_id <span class="sc">==</span> <span class="st">&quot;Xq&quot;</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fl">0.04</span></span></code></pre></div>
<p>The five vectors are combined into one data frame describing the
chromosome arm selection model:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>table_arm_selection_rates <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Arm_ID =</span> arm_id, <span class="at">Chromosome =</span> arm_chromosome, <span class="at">Bin_start =</span> arm_start, <span class="at">Bin_end =</span> arm_end, <span class="at">s_rate =</span> arm_s)</span></code></pre></div>
</div>
<div id="driver-gene-selection-model" class="section level3">
<h3>Driver gene selection model</h3>
<p>If <code>selection_model</code> = <code>driver-gene-selection</code>
or <code>chrom-arm-and-driver-gene-selection</code>, then cancer is
driven solely or partially by gains, losses and mutations of driver
genes.</p>
<p>Five vectors in combination establish the driver gene selection
rates. Given values at the same location in the vectors, the gene with
ID <code>gene_id</code>, which functions as <code>gene_role</code>
(either <code>TSG</code> for Tumor Suppressor Genes or
<code>ONCOGENE</code> for Oncogenes) at bin <code>gene_bin</code> on
chromosome <code>gene_chromosome</code>, has selection rate
<code>gene_s</code>. Similar to the chromosome arm selection model, here
the driver gene’s selection rate also ranges from 0 to infinity.
However, which CNA events are selective or deleterious for a gene
depends also on its role. For a Tumor Suppressor Gene: - Amplification
of the wild-type alleles is deleterious. - Deletion of the wild-type
alleles is advantageous. - Amplification and deletion of the mutant
alleles are neutral. - Mutation, i.e. a wild-type allele changes to
mutant, is advantageous. On the other hand, for an Oncogene: -
Amplification of the wild-type allele or the mutant allele is
advantageous. - Deletion of the wild-type allele or the mutant allele is
deleterious. - Mutation, i.e. a wild-type allele changes to mutant, is
advantageous. The selective increases or decreases resulting from
gain/loss or mutation of a gene (making the cell more advantageous or
deleterious) depend on how far its selection rate is from 1.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>driver_gene_list <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">&quot;HGSOC_driver_genes.xlsx&quot;</span>)</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>gene_id <span class="ot">&lt;-</span> driver_gene_list<span class="sc">$</span>Gene_ID</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>gene_role <span class="ot">&lt;-</span> driver_gene_list<span class="sc">$</span>Gene_role</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>gene_chromosome <span class="ot">&lt;-</span> driver_gene_list<span class="sc">$</span>Chromosome</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>gene_bin <span class="ot">&lt;-</span> <span class="fu">round</span>(driver_gene_list<span class="sc">$</span>Start <span class="sc">/</span> CN_bin_length)</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>gene_s <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="fu">length</span>(gene_id))</span></code></pre></div>
<p>The five vectors are combined into one data frame describing the
driver gene selection model:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a>table_gene_selection_rates <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Gene_ID =</span> gene_id, <span class="at">Gene_role =</span> gene_role, <span class="at">s_rate =</span> gene_s, <span class="at">Chromosome =</span> gene_chromosome, <span class="at">Bin =</span> gene_bin)</span></code></pre></div>
</div>
<div id="update-the-model-variables-with-the-selection-model" class="section level3">
<h3>Update the model variables with the selection model</h3>
<p>After the table(s) are finished, we can update
<code>model_variables</code> with the variables for the selection model.
Include only the tables required for the selected selection model:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>model_variables <span class="ot">&lt;-</span> <span class="fu">BUILD_driver_library</span>(</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>    <span class="at">model_variables =</span> model_variables,</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>    <span class="at">table_arm_selection_rates =</span> table_arm_selection_rates,</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>    <span class="at">table_gene_selection_rates =</span> table_gene_selection_rates</span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="initial-cell-population" class="section level2">
<h2>Initial cell population</h2>
<p>We now specify the initial cell population. Here we assume that the
initial state consists of one clone, where <code>cell_count</code> is
its cell count. The variable <code>CN_matrix</code> specifies the
clone’s CN profile, here assumed to be normal diploid female cells. The
variable <code>drivers</code> defines the existing driver mutations in
the clone, here assumed to be none.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>cell_count <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>CN_matrix <span class="ot">&lt;-</span> <span class="fu">BUILD_cn_normal_XX</span>(model_variables<span class="sc">$</span>cn_info)</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>drivers <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>model_variables <span class="ot">&lt;-</span> <span class="fu">BUILD_initial_population</span>(</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>    <span class="at">model_variables =</span> model_variables,</span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a>    <span class="at">cell_count =</span> cell_count,</span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a>    <span class="at">CN_matrix =</span> CN_matrix,</span>
<span id="cb22-8"><a href="#cb22-8" tabindex="-1"></a>    <span class="at">drivers =</span> drivers</span>
<span id="cb22-9"><a href="#cb22-9" tabindex="-1"></a>)</span></code></pre></div>
<p>If the initial cell population consists of multiple clones, the
function <code>BUILD_initial_population</code> can be repeated for each
distinct clone with different <code>cell_count</code>,
<code>CN_matrix</code> and <code>drivers</code>.</p>
</div>
<div id="save-the-input-files" class="section level2">
<h2>Save the input files</h2>
<p>The final step of model variable preparation is using the function
<code>SAVE_model_variables</code> to save the variables defined thus far
in csv files. The variable <code>model_name</code> is incorporated into
the csv files’ names, and the simulator will look up files starting with
this string to set up the simulations. Therefore, if different models
are simulated in the same folder, they should be saved with different
<code>model_name</code> so as not to overwrite older models.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a>model_name <span class="ot">&lt;-</span> <span class="st">&quot;EXPERIMENT&quot;</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="fu">SAVE_model_variables</span>(<span class="at">model_name =</span> model_name, <span class="at">model_variables =</span> model_variables)</span></code></pre></div>
</div>
</div>
<div id="run-the-simulator" class="section level1">
<h1>Run the simulator</h1>
<p>The function <code>simulator_full_program</code> produces simulations
according to the model variables listed in the csv files resulting from
the previous section.</p>
<p>There are several options for conditioning on the simulations or
requesting specific results: - <code>model</code>: Prefix of the csv
input files. This should be the same as <code>model_name</code> from the
previous section. - <code>n_simulations</code>: Number of simulations
requested. - <code>compute_parallel</code>: Simulations are produced
sequentially if <code>FALSE</code>, or in parallel if <code>TRUE</code>.
Default is <code>FALSE</code>. - <code>n_cores</code>: Number of
computer CPUs to use to produce simulations in parallel (therefore only
applied when <code>compute_parallel</code>=<code>TRUE</code>). If
<code>n_cores</code> is not specified when
<code>compute_parallel</code>=<code>TRUE</code>, the algorithm uses all
available CPU cores except one. - <code>seed</code>: Seed to produce
replicable stochastic simulations. Only applicable when
<code>compute_parallel</code>=<code>FALSE</code>. -
<code>report_progress</code>: The simulator prints updates of the
progress if set to <code>TRUE</code>. - <code>save_simulation</code>:
All output variables from each simulation are saved in separate
<code>rda</code> files if set to <code>TRUE</code>. Default is
<code>TRUE</code>. - <code>output_variables</code>: Vector of specific
output variables from each simulation to return in the list format in R.
Default is <code>NULL</code>, in which case no variable is returned. -
<code>folder_workplace</code>: Name of folder in the current directory
to store output files from the simulations. If the folder does not
exist, it will be created. Default is <code>NULL</code>, in which case
files are saved within the current directory. - <code>R_libPaths</code>:
Name of folder where required R libraries are stored. Default is
<code>NULL</code>, in which case the simulator assumes relevent R
libraries are automatically usable. - <code>n_clones_min</code>: Lower
limit for number of clones at the ending time point for simulations to
be accepted. Default is <code>0</code>. - <code>n_clones_max</code>:
Upper limit for number of clones at the ending time point for
simulations to be accepted. Default is <code>Inf</code>. -
<code>neutral_variations</code>: Whether neutral CNAs are simulated on
top of the selective CNAs. - <code>save_newick_tree</code>: If
<code>TRUE</code>, the phylogeny of sampled cells from each simulation
is saved in the Newick format. Default is <code>FALSE</code>. -
<code>save_cn_profile</code>: If <code>TRUE</code>, the CN profiles of
sampled cells from each simulation are saved in a csv file. Default is
<code>FALSE</code>. - <code>format_cn_profile</code>: If
<code>long</code> when <code>save_cn_profile</code>=<code>TRUE</code>,
the CN profiles are saved in the long format (subsequent cells are
appended by row). If <code>wide</code> when
<code>save_cn_profile</code>=<code>TRUE</code>, the CN profiles are
saved in the wide format (each cell occupies a column). If
<code>both</code> when <code>save_cn_profile</code>=<code>TRUE</code>,
the CN profiles are saved in both formats (resulting in two csv files
per simulation). Default is <code>long</code>. -
<code>save_cn_clones</code>: If <code>TRUE</code>, the assignment of
which cell belongs in which clone is saved in a csv file. Default is
<code>FALSE</code>. - <code>internal_nodes_cn_info</code>: If
<code>TRUE</code> when <code>save_cn_profile</code>=<code>TRUE</code>,
the CN profiles of both the sampled cells and the internal nodes of
their phylogeny are saved in the requested format(s). The internal node
labels correspond to those in the Newick tree if
<code>save_newick_tree</code>=<code>TRUE</code>. Default is
<code>FALSE</code>. - <code>model_readcount</code>: If
<code>TRUE</code>, noisy readcounts are simulated from each sampled
cell’s CN profile and saved in a csv file. Default is
<code>FALSE</code>. - <code>model_readcount_base</code>: If
<code>truth</code> when <code>model_readcount</code>=<code>TRUE</code>,
noisy readcounts are simulated from the CN profile without neutral
variations. If <code>neuvar</code> when
<code>model_readcount</code>=<code>TRUE</code>, noisy readcounts are
simulated from the CN profile with neutral variations. Both are
simulated and saved in separate csv files if
<code>model_readcount_base</code>=<code>all</code> when
<code>model_readcount</code>=<code>TRUE</code>. Default is
<code>all</code>. - <code>HMM</code>: If <code>TRUE</code>, the cell’s
CN profiles are inferred back from the simulated noisy readcounts using
HMMcopy. Default is <code>FALSE</code>. - <code>HMM_containner</code>:
Specifies which HMMcopy container is used when
<code>HMM</code>=<code>TRUE</code>. Two available options are
<code>singularity</code> and <code>docker</code>, which might have
different efficiency.</p>
<ul>
<li><code>pseudo_corrected_readcount</code>: if <code>TRUE</code>, the
pseudo-corrected readcounts (= cell’s HMM-inferred mean readcount X bin
GC level) are computed and saved in a csv file. The specific data used
depends on <code>model_readcount_base</code>. Default is
<code>FALSE</code>.</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">simulator_full_program</span>(</span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>    <span class="at">model =</span> model_name,</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>    <span class="at">n_simulations =</span> <span class="dv">10</span>,</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>    <span class="at">save_cn_profile =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a>    <span class="at">save_cn_clones =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a>    <span class="at">internal_nodes_cn_info =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a>    <span class="at">report_progress =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-8"><a href="#cb24-8" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">10</span>,</span>
<span id="cb24-9"><a href="#cb24-9" tabindex="-1"></a>    <span class="at">compute_parallel =</span> <span class="cn">TRUE</span></span>
<span id="cb24-10"><a href="#cb24-10" tabindex="-1"></a>)</span>
<span id="cb24-11"><a href="#cb24-11" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb24-12"><a href="#cb24-12" tabindex="-1"></a><span class="co">#&gt; Parallel cluster with 9 cores...</span></span></code></pre></div>
<p>A variety of plots from the simulation:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="co"># plot_all(</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="co">#     model = model_name,</span></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="co">#     n_simulations = 10,</span></span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a><span class="co">#     compute_parallel = TRUE</span></span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a><span class="co"># )</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
