```{r setup}
library(readxl)
library(data.table)
#-----------------------------------------------------------------------
devtools::install_github("chandrew314/CINner")
library(CINner)
```

# ===========PART 1: FIND SELECTION AND MISSEG RATES FOR NON-WGD SAMPLES

First, we set up the fixed parameters for the simulator.

```{r}
#---Each cell's lifespan is exponentially distributed with mean 30 days
cell_lifespan <- 30
#---Simulation starts at 0 years
T_0 <- list(0, "year")
#---Simulation ends at 80 years
T_end <- list(80, "year")
#---All cells at final time are sampled, sample ID is SA01
Table_sample <- data.frame(Sample_ID = c("SA01"), Cell_count = c(Inf), Age_sample = c(80))
#---Mode of selection
selection_model <- "WGD-chrom-arm-selection"
#---Probability of missegregation per homolog per cell division
prob_CN_missegregation <- 5e-5
#---Formula for probability of missegregation per cell division
formula_CN_missegregation <- "per_homolog:prob_CN_missegregation * ( (alpha_aneuploidy-1)*WGD_count + 1 )"
#---Formula for probability of arm-missegregation per cell division
formula_CN_chrom_arm_missegregation <- "per_homolog:prob_CN_chrom_arm_missegregation * ( (alpha_aneuploidy-1)*WGD_count + 1 )"
#---Conditions for cell viability: upper bound on (average) ploidy
bound_average_ploidy <- Inf
#---Conditions for cell viability: upper bound on bins with total CN = 0 (nullisomy)
bound_homozygosity <- 0
#---Conditions for cell viability: upper bound on bin CN
bound_maximum_CN <- Inf
#---Conditions for cell viability: upper bound on bin CN normalized by ploidy
bound_maximum_CN_normalized <- 2
#---Conditions for cell viability: upper bound on WGD count throughout cell's lineage
bound_WGD <- 1
#---Total cell population is expected to follow a logistic growth curve up to 10,000 cells
vec_time <- T_0[[1]]:T_end[[1]]
L <- 10000
t_0 <- 20
k <- 0.3
vec_cell_count <- L / (1 + exp(-k * (vec_time - t_0)))
table_population_dynamics <- cbind(vec_time, vec_cell_count)
#---Create model variable dataframe
model_variables_base <- BUILD_general_variables(
    cell_lifespan = cell_lifespan,
    T_0 = T_0, T_end = T_end,
    Table_sample = Table_sample,
    selection_model = selection_model,
    prob_CN_missegregation = prob_CN_missegregation,
    formula_CN_missegregation = formula_CN_missegregation,
    formula_CN_chrom_arm_missegregation = formula_CN_chrom_arm_missegregation,
    bound_maximum_CN = bound_maximum_CN,
    bound_average_ploidy = bound_average_ploidy,
    bound_homozygosity = bound_homozygosity,
    bound_maximum_CN_normalized = bound_maximum_CN_normalized,
    bound_WGD = bound_WGD,
    table_population_dynamics = table_population_dynamics
)
```

Then, five vectors are used to fit the chromosome arm selection rates. 
Given values at the same location in the vectors, `arm_s` is the selection rate for the arm with ID `arm_id`, which runs from starting bin `arm_start` to ending bin `arm_end` on chromosome `arm_chromosome`. 
The selection rate is initialized to be 1 for each chromosome arm in `arm_ID`.

```{r}
#---Initialize library of chromosome arms
arm_id <- c(paste(model_variables_base$cn_info$Chromosome, "p", sep = ""), paste(model_variables_base$cn_info$Chromosome, "q", sep = ""))
arm_chromosome <- rep(model_variables_base$cn_info$Chromosome, 2)
arm_start <- c(rep(1, length(model_variables_base$cn_info$Chromosome)), model_variables_base$cn_info$Centromere_location + 1)
arm_end <- c(model_variables_base$cn_info$Centromere_location, model_variables_base$cn_info$Bin_count)
arm_s <- rep(1, length(arm_id))
```

After initializing the selection rate for each chromosome arm, we can update `model_variables` with the variables for the selection model. Include only the tables required for the selected selection model:

```{r}
model_variables_base <- BUILD_driver_library(
    model_variables = model_variables_base,
    table_arm_selection_rates = data.frame(Arm_ID = arm_id, Chromosome = arm_chromosome, Bin_start = arm_start, Bin_end = arm_end, s_rate = arm_s)
)
```

## Initial cell population

We now specify the initial cell population. Here we assume that the initial state consists of one clone, where `cell_count` is its cell count. The variable `CN_matrix` specifies the clone's CN profile, here assumed to be normal diploid female cells. The variable `drivers` defines the existing driver mutations in the clone, here assumed to be none.
We use the function `CHECK_model_variables` to check the variables defined thus far in csv files.

```{r}
#---Initial population consists of 20 cells, each consisting of 22 diploid autosomes
cell_count <- 20
CN_matrix <- BUILD_cn_normal_autosomes(model_variables_base$cn_info)
drivers <- list()
model_variables_base <- BUILD_initial_population(
    model_variables = model_variables_base,
    cell_count = cell_count,
    CN_matrix = CN_matrix,
    drivers = drivers
)
#---Check model variable dataframe for compatibility
model_variables_base <- CHECK_model_variables(model_variables_base)
```

We then create a dataframe of the reference copy number coordinates for each chromosome.

```{r}
#---Create reference copy number coordinates
size_CN_block_DNA <- as.numeric(model_variables_base$general_variables$Value[which(model_variables_base$general_variables$Variable == "size_CN_block_DNA")])
vec_chr <- c()
vec_start <- c()
vec_end <- c()
vec_width <- c()
for (i in 1:nrow(model_variables_base$cn_info)) {
    Chromosome <- model_variables_base$cn_info$Chromosome[i]
    Bin_count <- model_variables_base$cn_info$Bin_count[i]
    vec_chr <- c(vec_chr, rep(Chromosome, Bin_count))
    vec_start <- c(vec_start, seq(1, Bin_count * size_CN_block_DNA, by = size_CN_block_DNA))
    vec_end <- c(vec_end, seq(size_CN_block_DNA, Bin_count * size_CN_block_DNA, by = size_CN_block_DNA))
    vec_width <- c(vec_width, rep(size_CN_block_DNA, Bin_count))
}
copynumber_coordinates <- data.frame(
    chr = vec_chr,
    start = vec_start,
    end = vec_end,
    width = vec_width
)
```

## ABC fitting for WGD rate and WGD-aneuploidy rate for each PCAWG cancer type

Next, we start on ABC fitting for WGD rate and WGD-aneuploidy rate.
First, we need to process the patient data by classifying it into WGD or non-WGD. We remove cancer types with less than 10 non-WGD samples.

```{r}
#---Input list of PCAWG sample ID's
copynumber_PCAWG <- read_excel(system.file("pcawg_specimen_histology_August2016_v9.xlsx", package = "CINner"))
#---Find list of PCAWG cancer types
PCAWG_cancer_types <- unique(copynumber_PCAWG$histology_abbreviation)
PCAWG_cancer_types <- PCAWG_cancer_types[which(!is.na(PCAWG_cancer_types))]
#---Input WGD status for each PCAWG sample
PCAWG_wgd <- read.table(system.file("consensus.20170218.purity.ploidy.txt", package = "CINner"), header = TRUE)
#---Find list of PCAWG sample ID's for each cancer type
PCAWG_cancer_type_sample_ids <- vector("list", length(PCAWG_cancer_types))
for (i in 1:length(PCAWG_cancer_types)) {
    sample_ids <- copynumber_PCAWG$tcga_sample_uuid[which(copynumber_PCAWG$histology_abbreviation == PCAWG_cancer_types[i])]
    for (j in 1:length(sample_ids)) {
        sample_id <- sample_ids[j]
        if (!file.exists(system.file(paste0("consensus.20170119.somatic.cna.annotated/", sample_id, ".consensus.20170119.somatic.cna.annotated.txt"), package = "CINner"))) next
        tmp <- read.table(system.file(paste0("consensus.20170119.somatic.cna.annotated/", sample_id, ".consensus.20170119.somatic.cna.annotated.txt"), package = "CINner"), header = TRUE)
        if (max(tmp$star, na.rm = TRUE) < 2) next
        PCAWG_cancer_type_sample_ids[[i]] <- c(PCAWG_cancer_type_sample_ids[[i]], sample_id)
    }
}
#---Find list of PCAWG sample ID's for each cancer type that are non-WGD
PCAWG_cancer_type_nonWGD_sample_ids <- vector("list", length(PCAWG_cancer_types))
for (i in 1:length(PCAWG_cancer_types)) {
    sample_ids <- PCAWG_cancer_type_sample_ids[[i]]
    sample_wgd <- PCAWG_wgd[which(PCAWG_wgd$samplename %in% sample_ids), ]
    sample_nonwgd_ids <- sample_wgd$samplename[which(sample_wgd$wgd_uncertain == FALSE & sample_wgd$wgd_status == "no_wgd")]
    PCAWG_cancer_type_nonWGD_sample_ids[[i]] <- sample_nonwgd_ids
}
#---Remove cancer types with less than 10 non-WGD samples
sample_sizes <- sapply(PCAWG_cancer_type_nonWGD_sample_ids, length)
PCAWG_cancer_types <- PCAWG_cancer_types[-which(sample_sizes < 10)]
PCAWG_cancer_type_sample_ids <- PCAWG_cancer_type_sample_ids[-which(sample_sizes < 10)]
PCAWG_cancer_type_nonWGD_sample_ids <- PCAWG_cancer_type_nonWGD_sample_ids[-which(sample_sizes < 10)]
```

For the remaining samples, we create their CN profiles.

```{r}
#---Create list of CN profiles for each sample in each PCAWG cancer type
copynumber_DATA_cancer_types <- vector("list", length(PCAWG_cancer_types))
for (i in 1:length(PCAWG_cancer_types)) {
    cancer_type <- PCAWG_cancer_types[i]
    cancer_type_sample_ids <- PCAWG_cancer_type_sample_ids[[i]]
    copynumber_DATA_ls <- vector("list", length(cancer_type_sample_ids))
    for (j in 1:length(cancer_type_sample_ids)) {
        sample_id <- cancer_type_sample_ids[j]
        tmp <- read.table(system.file(paste0("consensus.20170119.somatic.cna.annotated/", sample_id, ".consensus.20170119.somatic.cna.annotated.txt"), package = "CINner"), header = TRUE)
        if (any(is.na(tmp$star))) tmp <- tmp[-which(is.na(tmp))]
        tmp <- tmp[which(tmp$star >= 2), ]
        tmp$donor_unique_id <- sample_id
        copynumber_DATA_ls[[j]] <- tmp
    }
    copynumber_DATA_cancer_types[[i]] <- rbindlist(copynumber_DATA_ls)
}
```

Now we create the prior distributions and specify the variables we want to fit: probability of WGD and the WGD-aneuploidy rate.
The prior distribution of the probability fo WGD (`10^:prob_CN_whole_genome_duplication`) is uniformly distributed from -5 to -3.
The prior distribution of WGD-aneuploidy rate (`alpha_aneuploidy`) is uniformly distributed from 0 to 200.
We use the proportion of samples classified as whole-genome doubled (`WGD_proportion`) and the difference between the fraction of the genome altered in whole-genome duplicated and non whole-genome duplicated samples (`FGA_difference`) as our target statistics.

```{r}
#---Create list of parameters and ranges for their prior distributions
list_parameters <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(list_parameters) <- c("Variable", "Type", "Lower_bound", "Upper_bound")
list_parameters[nrow(list_parameters) + 1, ] <- c("10^:prob_CN_whole_genome_duplication", "CNA_probability", -5, -3)
#list_parameters[nrow(list_parameters) + 1, ] <- c("prob_CN_whole_genome_duplication", "CNA_probability", 1e-6, 1e-5)
list_parameters[nrow(list_parameters) + 1, ] <- c("alpha_aneuploidy", "Other", 0, 200)
#---Create list of target statistics
list_targets <- c("WGD_proportion", "FGA_difference")
```

Now we fit probability of WGD and the WGD-aneuploidy rate with ABC random forest for each cancer type in PCAWG.
We skip fitting for cancer types with no WGD samples and use our previously fitted missegregation and selection rates for each chromosome homolog.

```{r}
#---ABC fitting for each PCAWG cancer type
# for (i in 1:length(PCAWG_cancer_types)) {
for (i in 1) {
    cancer_type <- PCAWG_cancer_types[i]
    ####################################################################
    library_name <- paste0("ABC-BULK-ARM-WGD-", cancer_type, "-alpha")
    ####################################################################
    #---Get statistics for PCAWG cancer type
    copynumber_DATA <- copynumber_DATA_cancer_types[[i]]
    cancer_type_stats <- get_WGD_stats_from_data(
        copynumber_DATA = copynumber_DATA,
        DATA_wgd = PCAWG_wgd,
        copynumber_coordinates = copynumber_coordinates,
        cn_info = model_variables_base$cn_info,
        list_targets = list_targets
    )
    #---Skip fitting for cancer types with no WGD samples
    if (cancer_type_stats[which(list_targets == "WGD_proportion")] <= 0) {
        print(paste0(cancer_type, " has no WGD samples"))
        next
    }
    #---Input fitted misseg rates and selection rates for PCAWG cancer type
    filename <- paste0(cancer_type, "_fitted_parameters.csv")
    cancer_type_fitted_parameters <- read.csv(filename)
    model_variables_base_cancer_type <- bulk_arm_CN_assign_paras(
        model_variables = model_variables_base,
        parameter_IDs = cancer_type_fitted_parameters$Variable,
        parameters = cancer_type_fitted_parameters$Best_value
    )
    #   Recalibrate misseg rates from per-division to per-homolog
    loc_missegregation_rate <- which(model_variables_base_cancer_type$general_variables$Variable == "prob_CN_missegregation")
    tmp <- as.numeric(model_variables_base_cancer_type$general_variables$Value[loc_missegregation_rate])
    model_variables_base_cancer_type$general_variables$Value[loc_missegregation_rate] <-
        1 - (1 - tmp)^(1 / (2 * nrow(model_variables_base_cancer_type$cn_info)))
    loc_arm_missegregation_rate <- which(model_variables_base_cancer_type$general_variables$Variable == "prob_CN_chrom_arm_missegregation")
    tmp <- as.numeric(model_variables_base_cancer_type$general_variables$Value[loc_arm_missegregation_rate])
    model_variables_base_cancer_type$general_variables$Value[loc_arm_missegregation_rate] <-
        1 - (1 - tmp)^(1 / (2 * nrow(model_variables_base_cancer_type$cn_info)))
    #---Build simulation library for PCAWG cancer type
    cat(paste0("\n\n\nSIMULATION LIBRARY FOR ", cancer_type, "   [", i, "/", length(PCAWG_cancer_types), "]", "\n"))
    start_time <- Sys.time()
    library_bulk_arm_WGD(
        library_name = library_name,
        model_variables = model_variables_base_cancer_type,
        list_parameters = list_parameters,
        list_targets = list_targets,
        ################################################################
        ABC_simcount = 1000,
        n_samples = 50,
        ################################################################
        R_libPaths = R_libPaths
    )
    end_time <- Sys.time()
    print(end_time - start_time)
    #---ABC fitting for PCAWG cancer type
    cat(paste0("\n\n\nFITTING FOR ", cancer_type, "   [", i, "/", length(PCAWG_cancer_types), "]", "\n"))
    fitting_bulk_arm_WGD(
        library_name = library_name,
        model_name = cancer_type,
        model_variables = model_variables_base_cancer_type,
        stats_DATA = cancer_type_stats,
        list_parameters = list_parameters,
        list_targets = list_targets,
        folder_workplace = library_name,
        R_libPaths = R_libPaths
    )
}
```