```{r setup}
library(readxl)
library(data.table)
#-----------------------------------------------------------------------
devtools::install_github("chandrew314/CINner")
library(CINner)
```

First, set up the fiexd parameters for the simulator.

```{r}
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~Set up model parameters
#---Each cell's lifespan is exponentially distributed with mean 30 days
cell_lifespan <- 30
#---Simulation starts at 0 years
T_0 <- list(0, "year")
#---Simulation ends at 80 years
T_end <- list(80, "year")
#---All cells at final time are sampled, sample ID is SA01
Table_sample <- data.frame(Sample_ID = c("SA01"), Cell_count = c(Inf), Age_sample = c(80))
#---Mode of selection
selection_model <- "chrom-arm-selection"
#---Probability of missegregation per homolog per cell division
prob_CN_missegregation <- 5e-5
#---Formula for probability of missegregation per cell division
formula_CN_missegregation <- "per_division:prob_CN_missegregation"
#---Formula for probability of arm-missegregation per cell division
formula_CN_chrom_arm_missegregation <- "per_division:prob_CN_chrom_arm_missegregation"
#---Conditions for cell viability: upper bound on (average) ploidy
bound_average_ploidy <- Inf
#---Conditions for cell viability: upper bound on bins with total CN = 0 (nullisomy)
bound_homozygosity <- 0
#---Conditions for cell viability: upper bound on bin CN
bound_maximum_CN <- Inf
#---Conditions for cell viability: upper bound on bin CN normalized by ploidy
bound_maximum_CN_normalized <- 4
#---Conditions for cell viability: upper bound on WGD count throughout cell's lineage
bound_WGD <- 1
#---Total cell population is expected to follow a logistic growth curve up to 10,000 cells
vec_time <- T_0[[1]]:T_end[[1]]
L <- 10000
t_0 <- 20
k <- 0.3
vec_cell_count <- L / (1 + exp(-k * (vec_time - t_0)))
table_population_dynamics <- cbind(vec_time, vec_cell_count)
#---Create model variable dataframe
model_variables <- BUILD_general_variables(
    cell_lifespan = cell_lifespan,
    T_0 = T_0, T_end = T_end,
    Table_sample = Table_sample,
    selection_model = selection_model,
    prob_CN_missegregation = prob_CN_missegregation,
    formula_CN_missegregation = formula_CN_missegregation,
    formula_CN_chrom_arm_missegregation = formula_CN_chrom_arm_missegregation,
    bound_average_ploidy = bound_average_ploidy,
    bound_homozygosity = bound_homozygosity,
    bound_maximum_CN = bound_maximum_CN,
    bound_maximum_CN_normalized = bound_maximum_CN_normalized,
    bound_WGD = bound_WGD,
    table_population_dynamics = table_population_dynamics
)
```

Then, Five vectors are used to fit the chromosome arm selection rates. 
Given values at the same location in the vectors, `arm_s` is the selection rate for the arm with ID `arm_id`, which runs from starting bin `arm_start` to ending bin `arm_end` on chromosome `arm_chromosome`. 
The selection rate is initialized to be 1 for each chromosome arm in `arm_ID`.

```{r}
#---Initialize library of chromosome arms
arm_id <- c(paste(model_variables$cn_info$Chromosome, "p", sep = ""), paste(model_variables$cn_info$Chromosome, "q", sep = ""))
arm_chromosome <- rep(model_variables$cn_info$Chromosome, 2)
arm_start <- c(rep(1, length(model_variables$cn_info$Chromosome)), model_variables$cn_info$Centromere_location + 1)
arm_end <- c(model_variables$cn_info$Centromere_location, model_variables$cn_info$Bin_count)
arm_s <- rep(1, length(arm_id))
```

After initializing the selection rate for each chromosome arm, we can update `model_variables` with the variables for the selection model. Include only the tables required for the selected selection model:

```{r}
model_variables <- BUILD_driver_library(
    model_variables = model_variables,
    table_arm_selection_rates = data.frame(Arm_ID = arm_id, Chromosome = arm_chromosome, Bin_start = arm_start, Bin_end = arm_end, s_rate = arm_s)
)
```

## Initial cell population

We now specify the initial cell population. Here we assume that the initial state consists of one clone, where `cell_count` is its cell count. The variable `CN_matrix` specifies the clone's CN profile, here assumed to be normal diploid female cells. The variable `drivers` defines the existing driver mutations in the clone, here assumed to be none.

```{r}
#---Initial population consists of 20 cells, each consisting of 22 diploid autosomes
cell_count <- 20
CN_matrix <- BUILD_cn_normal_autosomes(model_variables$cn_info)
drivers <- list()
model_variables <- BUILD_initial_population(
    model_variables = model_variables,
    cell_count = cell_count,
    CN_matrix = CN_matrix,
    drivers = drivers
)
```

The final step of model variable preparation is using the function `CHECK_model_variables` to check the variables defined thus far in csv files. The variable `library_name` is incorporated into the csv files' names.

```{r}
#---Check model variable dataframe for compatibility
model_variables <- CHECK_model_variables(model_variables)
#---Set up name of library
library_name <- "ABC-BULK-ARM-CN"
```

## ABC Fitting for arm selection and missegregation rates
### Building the simulation library

`list_parameters_library` is a dataframe consisting of the parameters ('Variable') to be fitted: `prob_CN_chrom_arm_missegregation` and the selection rate for each chromosome arm. 
`prob_CN_chrom_arm_missegregation` has lower and upper bounds of 10^-5^ and 10^-4^, respectively. Each chromosome arm selection rate has lower and upper bounds of 0.5 and 1.5, respectively.

```{r}
#---Create list of parameters and ranges for their prior distributions
list_parameters_library <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(list_parameters_library) <- c("Variable", "Type", "Lower_bound", "Upper_bound")
list_parameters_library[nrow(list_parameters_library) + 1, ] <- c("prob_CN_chrom_arm_missegregation", "CNA_probability", 1e-5, 1e-4)
for (i in 1:length(model_variables$chromosome_arm_library$Arm_ID)) {
    list_parameters_library[nrow(list_parameters_library) + 1, ] <- c(
        model_variables$chromosome_arm_library$Arm_ID[i],
        "Arm_selection_rate",
        0.5, 1.5
    )
}
```
Initialize a vector specifying the variables we want to fit.
```{r}
#---Create list of target statistics
list_targets_library <- model_variables$chromosome_arm_library$Arm_ID
```

Here, we build a library of simulations using our specified model variables. 
Our variables to fit are specified by the `list_targets` variable. 
To generate statistics necessary for ABC, `n_sample` is the number of values from the prior distribution that are randomly selected for each simulation. 
`ABC_simcount` is then the number of simulations for a given parameters set. 
Each simulation is characterized by the copy number profile of the larest clone. 
The gain/loss frequencies of each chromosome arm are the important statistic used later for fitting arm selection rates.

```{r}
#---Build simulation library
start_time <- Sys.time()
library_bulk_arm_CN(
    library_name = library_name,
    model_variables = model_variables,
    list_parameters = list_parameters_library,
    list_targets = list_targets_library,
    ABC_simcount = 2000,
    n_samples = 100,
    R_libPaths = R_libPaths
)
end_time <- Sys.time()
print(end_time - start_time)
```

### ABC fitting for chromosome arm selection and missegregation

For fitting arm selection rates by cancer subtype in PCAWG data, we retrieve the different types of cancer present in PCAWG data and save it to `PCAWG_cancer_types`.

```{r}
#---Input list of PCAWG sample ID's
copynumber_PCAWG <- read_excel(system.file("pcawg_specimen_histology_August2016_v9.xlsx", package = "CINner"))
#---Find list of PCAWG cancer types
PCAWG_cancer_types <- unique(copynumber_PCAWG$histology_abbreviation)
PCAWG_cancer_types <- PCAWG_cancer_types[which(!is.na(PCAWG_cancer_types))]
```

We read whole genome duplication data from PCAWG and find the sample IDs associated with each cancer type.

```{r}
#---Input WGD status for each PCAWG sample
PCAWG_wgd <- read.table(system.file("consensus.20170218.purity.ploidy.txt", package = "CINner"), header = TRUE)
#---Find list of PCAWG sample ID's for each cancer type
PCAWG_cancer_type_sample_ids <- vector("list", length(PCAWG_cancer_types))
for (i in 1:length(PCAWG_cancer_types)) {
    sample_ids <- copynumber_PCAWG$tcga_sample_uuid[which(copynumber_PCAWG$histology_abbreviation == PCAWG_cancer_types[i])]
    for (j in 1:length(sample_ids)) {
        sample_id <- sample_ids[j]
        if (!file.exists(system.file(paste0("consensus.20170119.somatic.cna.annotated/", sample_id, ".consensus.20170119.somatic.cna.annotated.txt"), package = "CINner"))) next
        tmp <- read.table(system.file(paste0("consensus.20170119.somatic.cna.annotated/", sample_id, ".consensus.20170119.somatic.cna.annotated.txt"), package = "CINner"), header = TRUE)
        if (max(tmp$star, na.rm = TRUE) < 2) next
        PCAWG_cancer_type_sample_ids[[i]] <- c(PCAWG_cancer_type_sample_ids[[i]], sample_id)
    }
}
```

We include only the PCWAG cancer types that contain more than 10 samples with WGD.

```{r}
#---Find list of PCAWG sample ID's for each cancer type that are non-WGD
PCAWG_cancer_type_nonWGD_sample_ids <- vector("list", length(PCAWG_cancer_types))
for (i in 1:length(PCAWG_cancer_types)) {
    sample_ids <- PCAWG_cancer_type_sample_ids[[i]]
    sample_wgd <- PCAWG_wgd[which(PCAWG_wgd$samplename %in% sample_ids), ]
    sample_nonwgd_ids <- sample_wgd$samplename[which(sample_wgd$wgd_uncertain == FALSE & sample_wgd$wgd_status == "no_wgd")]
    PCAWG_cancer_type_nonWGD_sample_ids[[i]] <- sample_nonwgd_ids
}
#---Remove cancer types with less than 10 non-WGD samples
sample_sizes <- sapply(PCAWG_cancer_type_nonWGD_sample_ids, length)
PCAWG_cancer_types <- PCAWG_cancer_types[-which(sample_sizes < 10)]
PCAWG_cancer_type_sample_ids <- PCAWG_cancer_type_sample_ids[-which(sample_sizes < 10)]
PCAWG_cancer_type_nonWGD_sample_ids <- PCAWG_cancer_type_nonWGD_sample_ids[-which(sample_sizes < 10)]
#---Create list of CN profiles for each non-WGD sample in each PCAWG cancer type
copynumber_DATA_cancer_types <- vector("list", length(PCAWG_cancer_types))
for (i in 1:length(PCAWG_cancer_types)) {
    cancer_type <- PCAWG_cancer_types[i]
    cancer_type_sample_ids <- PCAWG_cancer_type_nonWGD_sample_ids[[i]]
    copynumber_DATA_ls <- vector("list", length(cancer_type_sample_ids))
    for (j in 1:length(cancer_type_sample_ids)) {
        sample_id <- cancer_type_sample_ids[j]
        tmp <- read.table(system.file(paste0("consensus.20170119.somatic.cna.annotated/", sample_id, ".consensus.20170119.somatic.cna.annotated.txt"), package = "CINner"), header = TRUE)
        if (any(is.na(tmp$star))) tmp <- tmp[-which(is.na(tmp))]
        tmp <- tmp[which(tmp$star >= 2), ]
        tmp$donor_unique_id <- sample_id
        copynumber_DATA_ls[[j]] <- tmp
    }
    copynumber_DATA_cancer_types[[i]] <- rbindlist(copynumber_DATA_ls)
}
```

Fit the chromosome arm missegregation and selection rates with ABC random forest for the remaining PCAWG cancer types.

```{r}
#---Find list of parameters
list_parameters <- list_parameters_library
#---Find list of target statistics
list_targets <- model_variables$chromosome_arm_library$Arm_ID
#---ABC fitting for each PCAWG cancer type
for (i in 1:length(PCAWG_cancer_types)) {
    cancer_type <- PCAWG_cancer_types[i]
    copynumber_DATA <- copynumber_DATA_cancer_types[[i]]
    cat(paste0("\n\n\nFITTING FOR ", cancer_type, "   [", i, "/", length(PCAWG_cancer_types), "]", "\n"))
    fitting_bulk_arm_CN(
        library_name = library_name,
        model_name = cancer_type,
        model_variables = model_variables,
        copynumber_DATA = copynumber_DATA,
        list_parameters = list_parameters,
        list_parameters_library = list_parameters_library,
        list_targets = list_targets,
        list_targets_library = list_targets_library,
        bound_freq = 0.1,
        library_shuffle = TRUE,
        folder_workplace = library_name,
        R_libPaths = R_libPaths
    )
}
```

Plot the results of the ABC fitting.

```{r}
#---Plot results from ABC fitting
df_WGD_FGA <- statistics_bulk_arm_WGD_status(
    plotname = "ALL_PCAWG",
    DATA_cancer_types = PCAWG_cancer_types,
    DATA_cancer_type_sample_ids = PCAWG_cancer_type_sample_ids,
    DATA_cancer_type_cn = copynumber_DATA_cancer_types,
    DATA_wgd = PCAWG_wgd,
    model_variables = model_variables
)
```