```{r setup}
library(readxl)
#-----------------------------------------------------------------------
setwd("/Users/dinhngockhanh/CancerSimulator/R/")
files_sources <- list.files(pattern = "*.r$")
sapply(files_sources, source)
#-----------------------------------------------------------------------
# devtools::install_github("dinhngockhanh/CancerSimulator")
# library(CancerSimulator)
```

#   Fit the lengths of focal events
.....
.....
.....
.....
.....
.....
.....

#   Get the Driver Gene data for a cancer type from PCAWG

We first load the list of samples for all cancer types in PCAWG:

```{r}
copynumber_PCAWG <- read_excel(system.file("pcawg_sample_sheet.xlsx", package = "CancerSimulator"))
```

List of cancer type-specific datasets in PCAWG:

```{r}
PCAWG_cancer_types <- unique(copynumber_PCAWG$dcc_project_code)
```

List of available samples per PCAWG cancer type:

```{r}
PCAWG_cancer_type_sample_ids <- vector("list", length(PCAWG_cancer_types))
for (i in 1:length(PCAWG_cancer_types)) {
    sample_ids <- copynumber_PCAWG$aliquot_id[which(copynumber_PCAWG$dcc_project_code == PCAWG_cancer_types[i])]
    for (j in 1:length(sample_ids)) {
        sample_id <- sample_ids[j]
        if (file.exists(system.file(paste0("consensus.20170119.somatic.cna.annotated/", sample_id, ".consensus.20170119.somatic.cna.annotated.txt"), package = "CancerSimulator"))) {
            tmp <- read.table(system.file(paste0("consensus.20170119.somatic.cna.annotated/", sample_id, ".consensus.20170119.somatic.cna.annotated.txt"), package = "CancerSimulator"), header = TRUE)
            if (max(tmp$star, na.rm = TRUE) < 2) next
            PCAWG_cancer_type_sample_ids[[i]] <- c(PCAWG_cancer_type_sample_ids[[i]], sample_id)
        }
    }
}
```

List of driver events (mutations and CNAs) per PCAWG sample:

```{r}
PCAWG_cancer_type_driver_events <- read.table(system.file("TableS3_panorama_driver_mutations_ICGC_samples.controlled.tsv", package = "CancerSimulator"), header = TRUE)
```

Decide which PCAWG cancer type to fit with the driver gene selection model:

```{r}
cancer_type <- "CLLE-ES"
cancer_type_sample_ids <- PCAWG_cancer_type_sample_ids[[which(PCAWG_cancer_types == cancer_type)]]
```

Get list of driver events for the PCAWG cancer type (including only mutational and CNA events, excluding germline and SV events)

```{r}
cancer_type_driver_events <- PCAWG_cancer_type_driver_events[
    which(PCAWG_cancer_type_driver_events$sample_id %in% cancer_type_sample_ids &
        PCAWG_cancer_type_driver_events$top_category %in% c("CNA", "mutational")),
]
```

Get list of driver genes for the PCAWG cancer type

```{r}
cancer_type_gene_id <- unique(cancer_type_driver_events$gene)
```

Get locations for the driver genes from Cancer Gene Census, and delete genes that are not in Cancer Gene Census or are on sex chromosomes

```{r}
CGC <- read.csv(system.file("extdata/cancer_gene_census.csv", package = "CancerSimulator"), header = TRUE)
cancer_type_gene_chromosome <- rep("", length(cancer_type_gene_id))
cancer_type_gene_pos <- rep(0, length(cancer_type_gene_id))
genes_to_delete <- c()
for (i in 1:length(cancer_type_gene_id)) {
    loc <- which(CGC$Gene.Symbol == cancer_type_gene_id[i])
    if (length(loc) == 0) {
        genes_to_delete <- c(genes_to_delete, cancer_type_gene_id[i])
        next
    }
    cancer_type_gene_chromosome[i] <- sub("\\:.*", "", CGC$Genome.Location[loc])
    cancer_type_gene_pos[i] <- strtoi(sub("\\-.*", "", sub(".*:", "", CGC$Genome.Location[loc])))
}
genes_to_delete <- c(genes_to_delete, cancer_type_gene_id[which(cancer_type_gene_chromosome %in% c("X", "Y"))])
if (length(genes_to_delete) > 0) {
    cancer_type_driver_events <- cancer_type_driver_events[-which(cancer_type_driver_events$gene %in% genes_to_delete), ]
    cancer_type_gene_chromosome <- cancer_type_gene_chromosome[-which(cancer_type_gene_id %in% genes_to_delete)]
    cancer_type_gene_pos <- cancer_type_gene_pos[-which(cancer_type_gene_id %in% genes_to_delete)]
    cancer_type_gene_id <- cancer_type_gene_id[-which(cancer_type_gene_id %in% genes_to_delete)]
}
```

Assign roles for the driver genes based on whether they are more frequently gained (ONCOGENE) or lost (TSG) in the PCAWG cancer type. If the genes are evenly divided between gains and lost or are not affected by CNAs, the gene roles come from Cancer Gene Census. If a gene is assigned as both ONCOGENE and TSG by Cancer Gene Census, then it is assigned TSG if it is not amplified in any sample in the PCAWG cancer type.

```{r}
cancer_type_gene_role <- rep("", length(cancer_type_gene_id))
for (i in 1:length(cancer_type_gene_id)) {
    freq_gain <- length(which(cancer_type_driver_events$gene == cancer_type_gene_id[i] & cancer_type_driver_events$category == "coding_amplification"))
    freq_loss <- length(which(cancer_type_driver_events$gene == cancer_type_gene_id[i] & cancer_type_driver_events$category == "coding_deletion"))
    if (freq_gain == freq_loss) {
        loc <- which(CGC$Gene.Symbol == cancer_type_gene_id[i])
        gene_role <- CGC$Role.in.Cancer[loc]
        if (grepl("oncogene", gene_role, fixed = TRUE) & grepl("TSG", gene_role, fixed = TRUE)) {
            if (freq_gain==0){
                cancer_type_gene_role[i] <- "TSG"
            } else {
                cancer_type_gene_role[i] <- "UNKNOWN"
            }
        } else if (grepl("oncogene", gene_role, fixed = TRUE)) {
            cancer_type_gene_role[i] <- "ONCOGENE"
        } else if (grepl("TSG", gene_role, fixed = TRUE)) {
            cancer_type_gene_role[i] <- "TSG"
        } else {
            cancer_type_gene_role[i] <- "UNKNOWN"
        }
    } else if (freq_gain > freq_loss) {
        cancer_type_gene_role[i] <- "ONCOGENE"
    } else {
        cancer_type_gene_role[i] <- "TSG"
    }
}
```

<!-- If the gene roles are still unknown, decide based on whether other genes on the chromosome are more ONCOGENE or TSG. If that still doesn't work, the roles are assigned at random.

```{r}
index_genes_unknown <- which(cancer_type_gene_role == "UNKNOWN")
for (i in 1:length(index_genes_unknown)) {
    other_gene_roles <- cancer_type_gene_role[which(cancer_type_gene_chromosome == cancer_type_gene_chromosome[index_genes_unknown[i]])]
    if (length(which(other_gene_roles == "TSG")) > length(which(other_gene_roles == "ONCOGENE"))) {
        cancer_type_gene_role[index_genes_unknown[i]] <- "TSG"
    } else if (length(which(other_gene_roles == "ONCOGENE")) > length(which(other_gene_roles == "TSG"))) {
        cancer_type_gene_role[index_genes_unknown[i]] <- "ONCOGENE"
    } else {
        if (runif(1) > 0.5) {
            cancer_type_gene_role[index_genes_unknown[i]] <- "ONCOGENE"
        } else {
            cancer_type_gene_role[index_genes_unknown[i]] <- "TSG"
        }
    }
}
``` -->

#   Create simulated gene-level library with CancerSimulator

First we set up the fixed parameters:

```{r}
cell_lifespan <- 30
T_0 <- list(0, "year")
T_end <- list(80, "year")
Table_sample <- data.frame(Sample_ID = c("SA01"), Cell_count = c(Inf), Age_sample = c(80))
T_tau_step <- cell_lifespan / 2
CN_bin_length <- 500000
selection_model <- "driver-gene-selection"
#------------------------------------------------------CNA PROBABILITIES
prob_CN_whole_genome_duplication <- 0e-4
prob_CN_missegregation <- 0e-5
prob_CN_chrom_arm_missegregation <- 0e-5
prob_CN_focal_amplification <- 0e-5
prob_CN_focal_deletion <- 0e-5
prob_CN_cnloh_interstitial <- 0e-5
prob_CN_cnloh_terminal <- 0e-5
model_CN_focal_amplification_length <- "beta"
model_CN_focal_deletion_length <- "beta"
prob_CN_focal_amplification_length_shape_1 <- 0.758304780825031
prob_CN_focal_amplification_length_shape_2 <- 5.33873409782625
prob_CN_focal_deletion_length_shape_1 <- 0.814054548726361
prob_CN_focal_deletion_length_shape_2 <- 6.16614890284825
prob_CN_cnloh_interstitial_length <- 0.005
prob_CN_cnloh_terminal_length <- 0.005
#--------------------------------------------DRIVER MUTATION PROBABILITY
rate_driver <- 1e-16
rate_passenger <- 1e-11
#---------------------------------------------------VIABILITY THRESHOLDS
# bound_driver <- Inf
bound_driver <- 5
bound_average_ploidy <- 6
bound_homozygosity <- 0
bound_maximum_CN <- Inf
bound_maximum_CN_normalized <- 4
#----------------------------------------------------POPULATION DYNAMICS
vec_time <- T_0[[1]]:T_end[[1]]
L <- 10000
t_0 <- 20
k <- 0.3
vec_cell_count <- L / (1 + exp(-k * (vec_time - t_0)))
table_population_dynamics <- cbind(vec_time, vec_cell_count)
#-----------------------------------------------------------------------
gc <- read.csv(file = system.file("extdata", "gc_map_500kb.csv", package = "CancerSimulator"))
gc_slope <- 1.2
gc_int <- 0
sigma1 <- 0.1
num_reads <- 2e6
model_variables <- BUILD_general_variables(
    cell_lifespan = cell_lifespan,
    T_0 = T_0, T_end = T_end, T_tau_step = T_tau_step,
    Table_sample = Table_sample,
    CN_bin_length = CN_bin_length,
    prob_CN_whole_genome_duplication = prob_CN_whole_genome_duplication,
    prob_CN_missegregation = prob_CN_missegregation,
    prob_CN_chrom_arm_missegregation = prob_CN_chrom_arm_missegregation,
    prob_CN_focal_amplification = prob_CN_focal_amplification,
    prob_CN_focal_deletion = prob_CN_focal_deletion,
    prob_CN_cnloh_interstitial = prob_CN_cnloh_interstitial,
    prob_CN_cnloh_terminal = prob_CN_cnloh_terminal,
    model_CN_focal_amplification_length = model_CN_focal_amplification_length,
    model_CN_focal_deletion_length = model_CN_focal_deletion_length,
    prob_CN_focal_amplification_length_shape_1 = prob_CN_focal_amplification_length_shape_1,
    prob_CN_focal_amplification_length_shape_2 = prob_CN_focal_amplification_length_shape_2,
    prob_CN_focal_deletion_length_shape_1 = prob_CN_focal_deletion_length_shape_1,
    prob_CN_focal_deletion_length_shape_2 = prob_CN_focal_deletion_length_shape_2,
    prob_CN_cnloh_interstitial_length = prob_CN_cnloh_interstitial_length,
    prob_CN_cnloh_terminal_length = prob_CN_cnloh_terminal_length,
    rate_driver = rate_driver,
    rate_passenger = rate_passenger,
    selection_model = selection_model,
    bound_driver = bound_driver,
    bound_maximum_CN = bound_maximum_CN,
    bound_average_ploidy = bound_average_ploidy,
    bound_homozygosity = bound_homozygosity,
    bound_maximum_CN_normalized = bound_maximum_CN_normalized,
    table_population_dynamics = table_population_dynamics,
    gc = gc,
    gc_slope = gc_slope,
    gc_int = gc_int,
    sigma1 = sigma1,
    num_reads = num_reads
)

gene_id <- cancer_type_gene_id
gene_role <- cancer_type_gene_role
gene_chromosome <- cancer_type_gene_chromosome
gene_bin <- round(cancer_type_gene_pos / CN_bin_length)
gene_s <- rep(1, length(gene_id))

model_variables <- BUILD_driver_library(
    model_variables = model_variables,
    table_gene_selection_rates = data.frame(Gene_ID = gene_id, Gene_role = gene_role, s_rate = gene_s, Chromosome = gene_chromosome, Bin = gene_bin)
)

cell_count <- 20
CN_matrix <- BUILD_cn_normal_autosomes(model_variables$cn_info)
drivers <- list()
model_variables <- BUILD_initial_population(model_variables = model_variables, cell_count = cell_count, CN_matrix = CN_matrix, drivers = drivers)
library_name <- paste0("ABC-BULK-DRIVER-GENE-", cancer_type)
model_variables <- CHECK_model_variables(model_variables)
```

We create a list of all parameters to be fitted, their types (CNA probabilities or driver-gene selection rates), and the bounds for their respective uniform prior distributions:

```{r}
list_parameters <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(list_parameters) <- c("Variable", "Type", "Lower_bound", "Upper_bound")
list_parameters[nrow(list_parameters) + 1, ] <- c("prob_CN_focal_amplification", "CNA_probability", 1e-6, 1e-4)
list_parameters[nrow(list_parameters) + 1, ] <- c("prob_CN_focal_deletion", "CNA_probability", 1e-6, 1e-4)
for (i in 1:length(model_variables$driver_library$Gene_ID)) {
    list_parameters[nrow(list_parameters) + 1, ] <- c(
        model_variables$driver_library$Gene_ID[i],
        "Gene_selection_rate",
        1, 1.5
    )
}
```

We create a list for the statistics targets. This would be the list of driver genes. The mutation/amplification/deletion frequencies of each gene form the statistics for ABC.

```{r}
list_targets <- model_variables$driver_library$Gene_ID
```

We then create a simulated gene-level library with CancerSimulator:

```{r}
start_time <- Sys.time()
library_bulk_gene(
    library_name = library_name,
    model_variables = model_variables,
    list_parameters = list_parameters,
    list_targets = list_targets,
    ABC_simcount = 1000,
    n_samples = 100
)
end_time <- Sys.time()
print(end_time - start_time)
```

#   Fit the Driver Gene data for a cancer type from PCAWG

```{r}
fitting_bulk_gene(
    library_name = library_name,
    model_name = cancer_type,
    sample_ids_DATA = cancer_type_sample_ids,
    driver_events_DATA = cancer_type_driver_events,
    list_parameters = list_parameters,
    list_targets = list_targets,
    folder_workplace = library_name
)
```






```{r}

```
