```{r setup}
library(ggplot2)
library(ggtree)
library(signals)
library(dendextend)
library(fishplot)
library(ctc)
library(adephylo)
library(data.table)
library(HMMcopy)
library(plyr)
library(matrixStats)
library(parallel)
library(RColorBrewer)
library(ape)

library(gridExtra)

library(readxl)
library(phyloTop)
library(pbapply)
#-----------------------------------------------------------------------
# setwd("/Users/dinhngockhanh/CancerSimulator/R/")
# files_sources <- list.files(pattern = "*.r$")
# sapply(files_sources, source)
#-----------------------------------------------------------------------
devtools::install_github("dinhngockhanh/CancerSimulator")
library(CancerSimulator)
```

#   Get the gain-loss map from PCAWG

We load the Copy Number data from the OV-AU cohort from PCAWG:

```{r}
copynumber_PCAWG <- read.csv(file = system.file("extdata", "PCAWG_OV-AU.csv", package = "CancerSimulator"))
```

#   Get the gain-loss map from simulations

First we set up the parameters:

```{r}
cell_lifespan                       <- 30
T_0                                 <- list(0,'year')
T_end                               <- list(80,'year')
Table_sample <- data.frame(Sample_ID=c('SA01'),Cell_count=c(500),Age_sample=c(80))
T_tau_step                          <- cell_lifespan/2
CN_bin_length                       <- 500000
########################################################################
#######################################   1st block: event probabilities
########################################################################
# selection_model                     <- 'old'
selection_model                     <- 'chrom-arm-selection'

prob_CN_whole_genome_duplication    <- 0*1e-4

prob_CN_missegregation              <- 0e-4
prob_CN_chrom_arm_missegregation    <- 10e-4
# prob_CN_missegregation              <- 2e-4
# prob_CN_chrom_arm_missegregation    <- 8e-4
# prob_CN_missegregation              <- 4e-4
# prob_CN_chrom_arm_missegregation    <- 4e-4

prob_CN_focal_amplification         <- 0*1e-4
prob_CN_focal_deletion              <- 0*1e-4
prob_CN_cnloh_interstitial          <- 0*1e-4
prob_CN_cnloh_terminal              <- 0*1e-4

# prob_CN_whole_genome_duplication    <- 0.5*0.5e-4
# prob_CN_missegregation              <- 0.5*2e-4
# prob_CN_chrom_arm_missegregation    <- 0.5*2e-4
# prob_CN_focal_amplification         <- 0.5*1e-4
# prob_CN_focal_deletion              <- 0.5*2e-4
# prob_CN_cnloh_interstitial          <- 0.5*1e-4
# prob_CN_cnloh_terminal              <- 0.5*1e-4

# prob_CN_whole_genome_duplication    <- 0.5e-4
# prob_CN_missegregation              <- 2e-4
# prob_CN_chrom_arm_missegregation    <- 2e-4
# prob_CN_focal_amplification         <- 1e-4
# prob_CN_focal_deletion              <- 2e-4
# prob_CN_cnloh_interstitial          <- 1e-4
# prob_CN_cnloh_terminal              <- 1e-4

prob_CN_focal_amplification_length  <- 0.01
prob_CN_focal_deletion_length       <- 0.01
prob_CN_cnloh_interstitial_length   <- 0.01
prob_CN_cnloh_terminal_length       <- 0.01

rate_driver                         <- 0
# rate_driver                         <- 8e-16
########################################################################
########################################################################
########################################################################
rate_passenger                      <- 1e-11
bound_driver                        <- 3
bound_average_ploidy                <- 4.5
bound_homozygosity                  <- 500
vec_time                            <- T_0[[1]]:T_end[[1]]
L                                   <- 10000
t_0                                 <- 20
k                                   <- 0.3
vec_cell_count                      <- L/(1+exp(-k*(vec_time-t_0)))
table_population_dynamics           <- cbind(vec_time,vec_cell_count)
gc                                  <- read.csv(file = system.file("extdata", "gc_map_500kb.csv", package = "CancerSimulator"))
gc_slope                            <- 1.2
gc_int                              <- 0
sigma1                              <- 0.1
num_reads                           <- 2e6
model_variables   <- BUILD_general_variables(
    cell_lifespan=cell_lifespan,
    T_0=T_0,T_end=T_end,T_tau_step=T_tau_step,
    Table_sample=Table_sample,
    CN_bin_length=CN_bin_length,
    prob_CN_whole_genome_duplication=prob_CN_whole_genome_duplication,
    prob_CN_missegregation=prob_CN_missegregation,
    prob_CN_chrom_arm_missegregation=prob_CN_chrom_arm_missegregation,
    prob_CN_focal_amplification=prob_CN_focal_amplification,
    prob_CN_focal_deletion=prob_CN_focal_deletion,
    prob_CN_cnloh_interstitial=prob_CN_cnloh_interstitial,
    prob_CN_cnloh_terminal=prob_CN_cnloh_terminal,
    prob_CN_focal_amplification_length=prob_CN_focal_amplification_length,
    prob_CN_focal_deletion_length=prob_CN_focal_deletion_length,
    prob_CN_cnloh_interstitial_length=prob_CN_cnloh_interstitial_length,
    prob_CN_cnloh_terminal_length=prob_CN_cnloh_terminal_length,
    rate_driver=rate_driver,
    rate_passenger=rate_passenger,
    selection_model=selection_model,
    bound_driver=bound_driver,
    bound_average_ploidy=bound_average_ploidy,
    bound_homozygosity=bound_homozygosity,
    table_population_dynamics=table_population_dynamics,
    gc=gc,
    gc_slope=gc_slope,
    gc_int=gc_int,
    sigma1=sigma1,
    num_reads=num_reads)
########################################################################
##########################   2nd block: driver genes and selection rates
########################################################################



#---------------------------For selection_model <- 'chrom-arm-selection'
vec_id              <- c(paste(model_variables$cn_info$Chromosome,'p',sep=''), paste(model_variables$cn_info$Chromosome,'q',sep=''))
vec_chromosome      <- rep(model_variables$cn_info$Chromosome,2)
vec_start           <- c(rep(1,length(model_variables$cn_info$Chromosome)), model_variables$cn_info$Centromere_location+1)
vec_end             <- c(model_variables$cn_info$Centromere_location, model_variables$cn_info$Bin_count)



vec_arm_s           <- rep(1,length(vec_id))
# vec_arm_s[which(vec_id=='1p')]  <- 1+0.03
vec_arm_s[which(vec_id=='5p')]  <- 1+0.3
vec_arm_s[which(vec_id=='5q')]  <- 1-0.3


model_variables   <- BUILD_driver_library(
    model_variables=model_variables,
    vec_id=vec_id,
    vec_chromosome=vec_chromosome,
    vec_start=vec_start,
    vec_end=vec_end,
    vec_arm_s=vec_arm_s
)



#------------------------------For selection_model <- 'old' or 'ancient'
# driver_gene_list    <- read_excel("HGSOC_driver_genes.xlsx")
# vec_driver_genes    <- driver_gene_list$Gene_ID
# vec_driver_role     <- driver_gene_list$Gene_role
# vec_chromosome      <- driver_gene_list$Chromosome
# vec_bin             <- round(driver_gene_list$Start/CN_bin_length)
# vec_driver_s        <- rep(0.003,length(vec_driver_genes))
# vec_driver_s[which(vec_driver_genes=='TP53')]   <- 0.01
# vec_driver_s[which(vec_driver_genes=='BRCA1')]  <- 0.005
# vec_driver_s[which(vec_driver_genes=='BRCA2')]  <- 0.005
# model_variables   <- BUILD_driver_library(
#     model_variables=model_variables,
#     vec_driver_genes=vec_driver_genes,
#     vec_driver_role=vec_driver_role,
#     vec_chromosome=vec_chromosome,
#     vec_bin=vec_bin,
#     vec_driver_s=vec_driver_s
# )
########################################################################
########################################################################
########################################################################
cell_count        <- 20
CN_matrix         <- BUILD_cn_normal_XX(model_variables$cn_info)
drivers           <- list()
model_variables   <- BUILD_initial_population(model_variables=model_variables,cell_count=cell_count,CN_matrix=CN_matrix,drivers=drivers)
model_name        <- 'HGSOC'
SAVE_model_variables(model_name=model_name,model_variables=model_variables)
```

Then create the same number of simulations as the sample count in PCAWG cohort:

```{r}
stage_final         <- 3
n_clones_min        <- 1
n_clones_max        <- Inf
n_simulations       <- length(unique(copynumber_PCAWG$donor_unique_id))
start_time = Sys.time()
simulations     <- simulator_full_program(
    model = model_name,
    n_simulations = n_simulations,
    stage_final = stage_final,
    n_clones_min = n_clones_min,
    n_clones_max = n_clones_max,
    save_simulation = FALSE,
    compute_parallel = TRUE,
    report_progress = TRUE,
    output_variables = c(
        'all_sample_genotype',
        'sample_cell_ID',
        'sample_genotype_unique',
        'sample_genotype_unique_profile',
        'cell_phylogeny_hclust'
    )
)
end_time = Sys.time()
print(end_time - start_time)
```

Combine all simulations into one variable:

```{r}
# numCores <- detectCores()
# cl <- makePSOCKcluster(numCores - 1)
# clusterExport(cl, varlist = c(
#     "model_name", "as.phylo"
# ))
# simulations <- parLapply(cl, 1:n_simulations, function(iteration) {
#     load(file=paste(model_name, "_simulation_", iteration, ".rda", sep = ""))
#     simulation_lite<-list()
#     simulation_lite$all_sample_genotype<-simulation$sample$all_sample_genotype
#     simulation_lite$sample_cell_ID<-simulation$sample$sample_cell_ID
#     simulation_lite$sample_genotype_unique<-simulation$sample$sample_genotype_unique
#     simulation_lite$sample_genotype_unique_profile<-simulation$sample$sample_genotype_unique_profile
#
#     simulation_lite$cell_phylogeny<-as.phylo(simulation$sample_phylogeny$cell_phylogeny_hclust)
#     return(simulation_lite)
# })
# stopCluster(cl)
```

#   Comparison between PCAWG and simulations

Plot the comparison between gain-loss maps from simulations and PCAWG:

```{r}
plot_gainloss(simulations,copynumber_PCAWG)
```

#   Comparison between DLP and simulations

We compare the statistics for cell phylogenies between the simulations and DLP data:

```{r}
phylo_stats_data <- read.csv(file = system.file("extdata", "dlp_phylo_topology.csv", package = "CancerSimulator"))
plot_phylo_stats(simulations,phylo_stats_data)
```


```{r}
```
