---
title: "How to use Cancer Simulator"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{How to use Cancer Simulator}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  warning = FALSE, message = FALSE,
  comment = "#>"
)
```

```{r setup}
library(pbapply)
library(ggimage)
library(ggplot2)
library(ggtree)
library(signals)
library(dendextend)
library(fishplot)
library(ctc)
library(adephylo)
library(data.table)
library(HMMcopy)
library(plyr)
library(matrixStats)
library(parallel)
library(RColorBrewer)
library(ape)

library(readxl)

# library(abc)
library(gridExtra)
#-----------------------------------------------------------------------
setwd("/Users/dinhngockhanh/CancerSimulator/R/")
files_sources <- list.files(pattern = "*.r$")
sapply(files_sources, source)
#-----------------------------------------------------------------------
# devtools::install_github("dinhngockhanh/CancerSimulator")
# library(CancerSimulator)
```

# Analyze the noise in DLP+ data

Here we analyze the noise level in `ov2295_cell_cn.csv`, from https://zenodo.org/record/3445364#.Ytq9ji-B3Pw.

```{r}
# ov2295_cn       <<- read.csv(file = 'ov2295_cell_cn.csv')
# table_gc          <<- read.csv(file = system.file("extdata", "gc_map_500kb.csv", package = "CancerSimulator"))
# analysis_dlp_cn(CNbins=ov2295_cn,table_gc=table_gc,filename_prefix='OV2295')
```

# Create one simulation

Save parameters to build one simulation:

```{r}
cell_lifespan <- 30
T_0 <- list(0, "year")
T_end <- list(80, "year")
Table_sample <- data.frame(Sample_ID = c("SA01"), Cell_count = c(3), Age_sample = c(80))
# Table_sample <- data.frame(Sample_ID = c("SA01"), Cell_count = c(300), Age_sample = c(80))
T_tau_step <- cell_lifespan / 2
CN_bin_length <- 500000
########################################################################
########################################################################
########################################################################
selection_model <- "chrom-arm-and-driver-gene-selection"

prob_CN_whole_genome_duplication <- 0e-4
prob_CN_missegregation <- 1e-4
prob_CN_chrom_arm_missegregation <- 1e-4
prob_CN_focal_amplification <- 1e-4
prob_CN_focal_deletion <- 1e-4
prob_CN_cnloh_interstitial <- 0e-5
prob_CN_cnloh_terminal <- 0e-5

prob_neutral_CN_whole_genome_duplication <- 0e-4
prob_neutral_CN_missegregation <- 100*prob_CN_missegregation
prob_neutral_CN_chrom_arm_missegregation <- 100*prob_CN_chrom_arm_missegregation
prob_neutral_CN_focal_amplification <- 100*prob_CN_focal_amplification
prob_neutral_CN_focal_deletion <- 100*prob_CN_focal_deletion
prob_neutral_CN_cnloh_interstitial <- 100*prob_CN_cnloh_interstitial
prob_neutral_CN_cnloh_terminal <- 100*prob_CN_cnloh_terminal

model_CN_focal_amplification_length <- "beta"
model_CN_focal_deletion_length <- "beta"
prob_CN_focal_amplification_length_shape_1 <- 0.758304780825031
prob_CN_focal_amplification_length_shape_2 <- 5.33873409782625
prob_CN_focal_deletion_length_shape_1 <- 0.814054548726361
prob_CN_focal_deletion_length_shape_2 <- 6.16614890284825
prob_CN_cnloh_interstitial_length <- 0.005
prob_CN_cnloh_terminal_length <- 0.005
rate_driver <- 0
########################################################################
########################################################################
########################################################################
rate_passenger <- 1e-11
bound_driver <- 3
bound_average_ploidy <- 4.5
bound_homozygosity <- 0

vec_time <- T_0[[1]]:T_end[[1]]
L <- 10000
t_0 <- 20
k <- 0.3
vec_cell_count <- L / (1 + exp(-k * (vec_time - t_0)))
table_population_dynamics <- cbind(vec_time, vec_cell_count)

gc <- read.csv(file = system.file("extdata", "gc_map_500kb.csv", package = "CancerSimulator"))
gc_slope <- 1.2
gc_int <- 0
sigma1 <- 0.02642392
num_reads <- 3906632
model_variables <- BUILD_general_variables(
    cell_lifespan = cell_lifespan,
    T_0 = T_0, T_end = T_end, T_tau_step = T_tau_step,
    Table_sample = Table_sample,
    CN_bin_length = CN_bin_length,

    prob_CN_whole_genome_duplication = prob_CN_whole_genome_duplication,
    prob_CN_missegregation = prob_CN_missegregation,
    prob_CN_chrom_arm_missegregation = prob_CN_chrom_arm_missegregation,
    prob_CN_focal_amplification = prob_CN_focal_amplification,
    prob_CN_focal_deletion = prob_CN_focal_deletion,
    prob_CN_cnloh_interstitial = prob_CN_cnloh_interstitial,
    prob_CN_cnloh_terminal = prob_CN_cnloh_terminal,

    prob_neutral_CN_whole_genome_duplication = prob_neutral_CN_whole_genome_duplication,
    prob_neutral_CN_missegregation = prob_neutral_CN_missegregation,
    prob_neutral_CN_chrom_arm_missegregation = prob_neutral_CN_chrom_arm_missegregation,
    prob_neutral_CN_focal_amplification = prob_neutral_CN_focal_amplification,
    prob_neutral_CN_focal_deletion = prob_neutral_CN_focal_deletion,
    prob_neutral_CN_cnloh_interstitial = prob_neutral_CN_cnloh_interstitial,
    prob_neutral_CN_cnloh_terminal = prob_neutral_CN_cnloh_terminal,

    model_CN_focal_amplification_length = model_CN_focal_amplification_length,
    model_CN_focal_deletion_length = model_CN_focal_deletion_length,
    prob_CN_focal_amplification_length_shape_1 = prob_CN_focal_amplification_length_shape_1,
    prob_CN_focal_amplification_length_shape_2 = prob_CN_focal_amplification_length_shape_2,
    prob_CN_focal_deletion_length_shape_1 = prob_CN_focal_deletion_length_shape_1,
    prob_CN_focal_deletion_length_shape_2 = prob_CN_focal_deletion_length_shape_2,
    prob_CN_cnloh_interstitial_length = prob_CN_cnloh_interstitial_length,
    prob_CN_cnloh_terminal_length = prob_CN_cnloh_terminal_length,
    rate_driver = rate_driver,
    rate_passenger = rate_passenger,
    selection_model = selection_model,
    bound_driver = bound_driver,
    bound_average_ploidy = bound_average_ploidy,
    bound_homozygosity = bound_homozygosity,
    table_population_dynamics = table_population_dynamics,
    gc = gc,
    gc_slope = gc_slope,
    gc_int = gc_int,
    sigma1 = sigma1,
    num_reads = num_reads
)
########################################################################
########################################################################
########################################################################
arm_id <- c(paste(model_variables$cn_info$Chromosome, "p", sep = ""), paste(model_variables$cn_info$Chromosome, "q", sep = ""))
arm_chromosome <- rep(model_variables$cn_info$Chromosome, 2)
arm_start <- c(rep(1, length(model_variables$cn_info$Chromosome)), model_variables$cn_info$Centromere_location + 1)
arm_end <- c(model_variables$cn_info$Centromere_location, model_variables$cn_info$Bin_count)
arm_s <- rep(1, length(arm_id))
arm_s[which(arm_id == "1p")] <- 1 + 0.05
arm_s[which(arm_id == "1q")] <- 1 + 0.065
arm_s[which(arm_id == "2p")] <- 1 + 0.06
arm_s[which(arm_id == "2q")] <- 1 + 0.05
arm_s[which(arm_id == "3p")] <- 1 + 0.02
arm_s[which(arm_id == "3q")] <- 1 + 0.095
arm_s[which(arm_id == "4p")] <- 1 - 0.013
arm_s[which(arm_id == "4q")] <- 1 - 0.016
arm_s[which(arm_id == "5p")] <- 1 + 0.18
arm_s[which(arm_id == "5q")] <- 1 - 0.1
arm_s[which(arm_id == "6p")] <- 1 + 0.2
arm_s[which(arm_id == "6q")] <- 1 - 0.17
arm_s[which(arm_id == "7p")] <- 1 + 0.02
arm_s[which(arm_id == "7q")] <- 1 + 0.1
arm_s[which(arm_id == "8p")] <- 1 - 0.14
arm_s[which(arm_id == "8q")] <- 1 + 0.2
arm_s[which(arm_id == "9p")] <- 1 + 0.022
arm_s[which(arm_id == "9q")] <- 1 + 0.022
arm_s[which(arm_id == "10p")] <- 1 + 0.03
arm_s[which(arm_id == "10q")] <- 1 + 0.035
arm_s[which(arm_id == "11p")] <- 1 - 0.15
arm_s[which(arm_id == "11q")] <- 1 + 0.13
arm_s[which(arm_id == "12p")] <- 1 + 0.08
arm_s[which(arm_id == "12q")] <- 1 + 0.04
arm_s[which(arm_id == "13p")] <- 1 + 0.03
arm_s[which(arm_id == "13q")] <- 1 + 0.022
arm_s[which(arm_id == "14p")] <- 1 + 0.022
arm_s[which(arm_id == "14q")] <- 1 + 0.016
arm_s[which(arm_id == "15p")] <- 1 + 0.02
arm_s[which(arm_id == "15q")] <- 1 + 0.01
arm_s[which(arm_id == "16p")] <- 1 + 0.01
arm_s[which(arm_id == "16q")] <- 1 - 0.08
arm_s[which(arm_id == "17p")] <- 1 - 0.07
arm_s[which(arm_id == "17q")] <- 1 + 0.01
arm_s[which(arm_id == "18p")] <- 1 - 0.04
arm_s[which(arm_id == "18q")] <- 1 - 0.03
arm_s[which(arm_id == "19p")] <- 1 + 0.06
arm_s[which(arm_id == "19q")] <- 1 + 0.06
arm_s[which(arm_id == "20p")] <- 1 + 0.05
arm_s[which(arm_id == "20q")] <- 1 + 0.06
arm_s[which(arm_id == "21p")] <- 1 - 0.22
arm_s[which(arm_id == "21q")] <- 1 + 0.08
arm_s[which(arm_id == "22p")] <- 1 - 0.02
arm_s[which(arm_id == "22q")] <- 1 - 0.02
arm_s[which(arm_id == "Xp")] <- 1 - 0.04
arm_s[which(arm_id == "Xq")] <- 1 - 0.04

driver_gene_list <- read_excel("HGSOC_driver_genes.xlsx")
gene_id <- driver_gene_list$Gene_ID
gene_role <- driver_gene_list$Gene_role
gene_chromosome <- driver_gene_list$Chromosome
gene_bin <- round(driver_gene_list$Start / CN_bin_length)
gene_s <- rep(1, length(gene_id))

table_arm_selection_rates <- data.frame(Arm_ID = arm_id, Chromosome = arm_chromosome, Bin_start = arm_start, Bin_end = arm_end, s_rate = arm_s)
table_gene_selection_rates <- data.frame(Gene_ID = gene_id, Gene_role = gene_role, s_rate = gene_s, Chromosome = gene_chromosome, Bin = gene_bin)
model_variables <- BUILD_driver_library(
    model_variables = model_variables,
    table_arm_selection_rates = table_arm_selection_rates,
    table_gene_selection_rates = table_gene_selection_rates
)
########################################################################
########################################################################
########################################################################
cell_count <- 20
CN_matrix <- BUILD_cn_normal_XX(model_variables$cn_info)
drivers <- list()
model_variables <- BUILD_initial_population(
    model_variables = model_variables,
    cell_count = cell_count,
    CN_matrix = CN_matrix,
    drivers = drivers
)
model_name <- "HMMCOPY-STUDY"
SAVE_model_variables(model_name = model_name, model_variables = model_variables)
```

Create one simulation:

```{r}
start_time <- Sys.time()
n_simulations <- 3
# n_simulations <- 10
tmp <- simulator_full_program(
    model = model_name, n_simulations = n_simulations,
    neutral_variations = TRUE,
    model_readcount = TRUE,
    model_readcount_base = 'neuvar',
    HMM = TRUE,
    HMM_containner = 'docker',
    compute_parallel = FALSE,
    seed = 10
)
end_time <- Sys.time()
print(end_time - start_time)
```

Make plots for this "original" simulation:

```{r}
# plot_all(model = model_name,n_simulations = n_simulations)



plot_clonal_phylo(
    model = model_name,
    n_simulations = n_simulations,
    width = 2000,
    height = 2000
)

plot_cn_heatmap(
    model = model_name,
    n_simulations = n_simulations,
    plotcol = "total-copy",
    CN_data = "TRUTH",
    phylo = "TRUTH",
    width = 1000,
    height = 1000
)

plot_cn_heatmap(
    model = model_name,
    n_simulations = n_simulations,
    plotcol = "total-copy",
    CN_data = "NEUTRAL-VARIATIONS",
    phylo = "TRUTH",
    width = 1000,
    height = 1000
)

plot_cn_heatmap(
    model = model_name,
    n_simulations = n_simulations,
    plotcol = "total-copy",
    CN_data = "NEUTRAL-VARIATIONS",
    phylo = "UMAP",
    width = 1000,
    height = 1000
)

plot_cn_heatmap(
    model = model_name,
    n_simulations = n_simulations,
    plotcol = "total-copy",
    CN_data = "NEUTRAL-VARIATIONS-HMM",
    phylo = "TRUTH",
    width = 1000,
    height = 1000
)

plot_cn_heatmap(
    model = model_name,
    n_simulations = n_simulations,
    plotcol = "total-copy",
    CN_data = "NEUTRAL-VARIATIONS-HMM",
    phylo = "UMAP",
    width = 1000,
    height = 1000
)
```







<!-- # Create noisy readcounts with varying parameters

Rename the "original" simulation and its plots:

```{r}
file.rename(paste(model_name, "_simulation_1.rda", sep = ""), paste(model_name, "_simulation_original.rda", sep = ""))
file.rename(paste(model_name, "_sim1_clonal_fishplot.jpeg", sep = ""), paste(model_name, "_original_clonal_fishplot.jpeg", sep = ""))
file.rename(paste(model_name, "_sim1_clonal_phylo.jpeg", sep = ""), paste(model_name, "_original_clonal_phylo.jpeg", sep = ""))
file.rename(paste(model_name, "_sim1_CN_minor_cnTRUTH_phyloTRUTH.jpeg", sep = ""), paste(model_name, "_original_CN_minor.jpeg", sep = ""))
file.rename(paste(model_name, "_sim1_CN_total_cnTRUTH_phyloTRUTH.jpeg", sep = ""), paste(model_name, "_original_CN_total.jpeg", sep = ""))
file.rename(paste(model_name, "_sim1_tot_pop.jpeg", sep = ""), paste(model_name, "_original_tot_pop.jpeg", sep = ""))
```

Load the "original" simulation:

```{r}
load(paste(model_name, "_simulation_original.rda", sep = ""))
simulation_original <- simulation
```

Create a folder as workplace to store files containing noisy readcounts and HMMcopy results:

```{r}
dir.create(model_name)
```

Save data for GC content and mappability as wig files to be used during HMMcopy:

```{r}
SIMULATOR_VARIABLES_for_simulation(model_name)
p2_write_gc_map_as_wig(filename_gc = paste(model_name, "_gc.wig", sep = ""), filename_map = paste(model_name, "_map.wig", sep = ""))
```

Set up variables for the noisy readcount model:

```{r}
table_gc            <<- read.table(paste(model_name, "-input-gc.csv", sep = ""), header = TRUE, sep = ",")
gc_int              <<- 0
gc_slope            <<- 1
vec.sigma1          <- c(seq(1,9,by=2)*1e-2, rep(0.02642392,length=5))
vec.num_reads       <- c(rep(3906632,length=5), seq(1,9,by=2)*1e6)
n_trials            <- length(vec.sigma1)
```

Create noisy readcounts for each parameter set:

```{r}
ind                         <- 0
for (ind in 1:length(vec.sigma1)){
    sigma1                  <<- vec.sigma1[ind]
    num_reads               <<- vec.num_reads[ind]
    #   Create noisy readcounts
    simulation              <- p2_readcount_model(simulation_original, TRUE)
    simulation              <- p2_readcount_model_wide(simulation, TRUE)
    #   Save noisy readcounts
    filename_ind            <- paste(model_name, "_simulation_", ind,".rda", sep = "")
    save(simulation, file = filename_ind)
    sample_cell_ID          <- simulation$sample$sample_cell_ID
    noisy_cn_profiles_long  <- simulation$sample$noisy_cn_profiles_long
    noisy_cn_profiles_wide  <- simulation$sample$noisy_cn_profiles_wide
    filename                <- paste(model_name, "_noisy_cn_profiles_long_", ind, ".csv", sep = "")
    write.csv(noisy_cn_profiles_long, filename, row.names = FALSE)
    filename                <- paste(model_name, "_noisy_cn_profiles_wide_", ind, ".csv", sep = "")
    write.csv(noisy_cn_profiles_wide, filename, row.names = FALSE)
    for (cell in 1:length(sample_cell_ID)) {
        cell_ID             <- sample_cell_ID[cell]
        filename            <- paste(model_name, "/", model_name, "_noisy_cn_profiles_long_", ind, "_", cell_ID, ".wig", sep = "")
        p2_write_cn_as_wig(filename, noisy_cn_profiles_long, cell_ID)
    }
}
```

# Run HMMcopy on noisy readcounts

Run file `hmmcopy.bash` which runs HMMcopy for each parameter set:

```{r}
#===========>   RUN ON GINSBURG:
#   Step 1:     Copy hmmcopy_v0.0.45.sif to folder HMMCOPY-STUDY (or whatever workplace)
#   Step 2:     Change model_name & n_simulations in ginsburg_hmmcopy.sh
#   Step 2:     Copy folder vignettes to Ginsburg/Terremoto
#   Step 3:     sbatch ginsburg_hmmcopy.sh
#===========>   HOW hmmcopy_v0.0.45.sif WAS CREATED:
#   On Ginsburg/Terremoto:
#               module load singularity
#               singularity pull docker://quay.io/mondrianscwgs/hmmcopy:v0.0.45





#===========    IF RUNNING ON COMPUTER: (EXTREMELY SLOW ON MAC M1)
# start_time = Sys.time()
# n_trials            <- length(vec.gc_slope)*length(vec.sigma1)
# flag_parallel       <- 1
# system("chmod -x hmmcopy.bash")
# system(paste("bash hmmcopy.bash ", model_name, " ", n_trials, " ", flag_parallel, sep = ""))
# end_time = Sys.time()
# print(end_time - start_time)
```

Incorporate HMMcopy results back into simulation files:

```{r}
append_with_hmm(model = model_name, n_simulations = n_trials, UMAP = TRUE, pseudo_corrected_readcount = FALSE)
```

Plot the GC-readcount plots:

```{r}
plot_gc_readcount(
    model = model_name,
    n_simulations = n_trials,
    width = 1000,
    height = 1000
)
```

Plot the total CN profiles inferred from HMMcopy:

```{r}
plot_cn_heatmap(
    model = model_name,
    n_simulations = n_trials,
    plotcol = "total-copy",
    CN_data = "HMM",
    phylo = "TRUTH",
    width = 1000,
    height = 1000
)

plot_cn_heatmap(
    model = model_name,
    n_simulations = n_trials,
    plotcol = "total-copy",
    CN_data = "HMM",
    phylo = "UMAP",
    width = 1000,
    height = 1000
)
```

Change filenames into something more recognizable:

```{r}
for (ind in 1:length(vec.sigma1)){
    sigma1                  <- vec.sigma1[ind]
    num_reads               <- vec.num_reads[ind]
    filename_old            <- paste(model_name,'_sim',ind,'_reads_vs_GC.jpeg',sep='')
    filename_new            <- paste('sig=',sigma1,'_cov=',num_reads,'_reads_vs_GC.jpeg',sep='')
    file.rename(filename_old, filename_new)
    filename_old            <- paste(model_name,'_sim',ind,'_CN_total_cnHMM_phyloTRUTH.jpeg',sep='')
    filename_new            <- paste('sig=',sigma1,'_cov=',num_reads,'_cn=HMM_phylo=TRUTH.jpeg',sep='')
    file.rename(filename_old, filename_new)
    filename_old            <- paste(model_name,'_sim',ind,'_CN_total_cnHMM_phyloUMAP.jpeg',sep='')
    filename_new            <- paste('sig=',sigma1,'_cov=',num_reads,'_cn=HMM_phylo=UMAP.jpeg',sep='')
    file.rename(filename_old, filename_new)
}
```

Find statistics for inferred CN from HMMcopy compared to true CN:

```{r}
hmm_stats   <- statistics_hmm(model = model_name, n_simulations = n_trials)
```

Plot the distribution of cells with right & wrong inferred ploidy from HMMcopy:

```{r}
filename <- paste(model_name, "_inferred_ploidy_in_HMM_var=sigma.jpeg", sep = "")
jpeg(file = filename, width = 2000, height = 1000)
df <- data.frame(
    Sigma = rep(vec.sigma1[1:5], 2),
    Percentage = c(
        100 * hmm_stats$ploidy_small[1:5] / (hmm_stats$ploidy_small[1:5] + hmm_stats$ploidy_right[1:5] + hmm_stats$ploidy_big[1:5]),
        100 * hmm_stats$ploidy_big[1:5] / (hmm_stats$ploidy_small[1:5] + hmm_stats$ploidy_right[1:5] + hmm_stats$ploidy_big[1:5])
    ),
    Group = c(rep("Inferred ploidy is too small", 5), rep("Inferred ploidy is too big", 5))
)
df$Sigma <- factor(df$Sigma, levels = vec.sigma1[1:5])
p <- ggplot(
    data = df,
    aes(x = Sigma, y = Percentage, fill = Group)
) +
    geom_bar(stat = "identity", position = position_dodge()) +
    theme(panel.background = element_rect(fill = "white", colour = "grey50")) +
    theme(text = element_text(size = 40), legend.position = "bottom")
print(p)
dev.off()

filename <- paste(model_name, "_inferred_cn_in_HMM_var=sigma.jpeg", sep = "")
jpeg(file = filename, width = 2000, height = 1000)
df <- data.frame(
    Sigma = rep(vec.sigma1[1:5], 2),
    Percentage = c(
        100 * hmm_stats$cn_small[1:5] / (hmm_stats$cn_small[1:5] + hmm_stats$cn_right[1:5] + hmm_stats$cn_big[1:5]),
        100 * hmm_stats$cn_big[1:5] / (hmm_stats$cn_small[1:5] + hmm_stats$cn_right[1:5] + hmm_stats$cn_big[1:5])
    ),
    Group = c(rep("Inferred CN is too small", 5), rep("Inferred CN is too big", 5))
)
df$Sigma <- factor(df$Sigma, levels = vec.sigma1[1:5])
p <- ggplot(
    data = df,
    aes(x = Sigma, y = Percentage, fill = Group)
) +
    geom_bar(stat = "identity", position = position_dodge()) +
    theme(panel.background = element_rect(fill = "white", colour = "grey50")) +
    theme(text = element_text(size = 40), legend.position = "bottom")
print(p)
dev.off()

filename <- paste(model_name, "_inferred_ploidy_in_HMM_var=coverage.jpeg", sep = "")
jpeg(file = filename, width = 2000, height = 1000)
df <- data.frame(
    Coverage = rep(vec.num_reads[6:10], 2),
    Percentage = c(
        100 * hmm_stats$ploidy_small[6:10] / (hmm_stats$ploidy_small[6:10] + hmm_stats$ploidy_right[6:10] + hmm_stats$ploidy_big[6:10]),
        100 * hmm_stats$ploidy_big[6:10] / (hmm_stats$ploidy_small[6:10] + hmm_stats$ploidy_right[6:10] + hmm_stats$ploidy_big[6:10])
    ),
    Group = c(rep("Inferred ploidy is too small", 5), rep("Inferred ploidy is too big", 5))
)
df$Coverage <- factor(df$Coverage, levels = vec.num_reads[6:10])
p <- ggplot(
    data = df,
    aes(x = Coverage, y = Percentage, fill = Group)
) +
    geom_bar(stat = "identity", position = position_dodge()) +
    theme(panel.background = element_rect(fill = "white", colour = "grey50")) +
    theme(text = element_text(size = 40), legend.position = "bottom")
print(p)
dev.off()

filename <- paste(model_name, "_inferred_cn_in_HMM_var=coverage.jpeg", sep = "")
jpeg(file = filename, width = 2000, height = 1000)
df <- data.frame(
    Coverage = rep(vec.num_reads[6:10], 2),
    Percentage = c(
        100 * hmm_stats$cn_small[6:10] / (hmm_stats$cn_small[6:10] + hmm_stats$cn_right[6:10] + hmm_stats$cn_big[6:10]),
        100 * hmm_stats$cn_big[6:10] / (hmm_stats$cn_small[6:10] + hmm_stats$cn_right[6:10] + hmm_stats$cn_big[6:10])
    ),
    Group = c(rep("Inferred CN is too small", 5), rep("Inferred CN is too big", 5))
)
df$Coverage <- factor(df$Coverage, levels = vec.num_reads[6:10])
p <- ggplot(
    data = df,
    aes(x = Coverage, y = Percentage, fill = Group)
) +
    geom_bar(stat = "identity", position = position_dodge()) +
    theme(panel.background = element_rect(fill = "white", colour = "grey50")) +
    theme(text = element_text(size = 40), legend.position = "bottom")
print(p)
dev.off()
``` -->


```{r}
```
