```{r setup}
library(readxl)
#-----------------------------------------------------------------------
setwd("/Users/dinhngockhanh/CancerSimulator/R/")
files_sources <- list.files(pattern = "*.r$")
sapply(files_sources, source)
#-----------------------------------------------------------------------
# devtools::install_github("dinhngockhanh/CancerSimulator")
# library(CancerSimulator)
```

First, we set up the fixed parameters for the simulator.
Note that we need to adjust the probability of chromosome miseggregation per division to reflect the fact that WGD cells have more chromosomes.
We set the 'rate' of aneuploidy to be 1 initially, which means that the rates of chromosome and chromosome arm miseggregation are the same between WGD and non-WGD cells.

```{r}
cell_lifespan <- 30
T_0 <- list(0, "year")
T_end <- list(80, "year")
Table_sample <- data.frame(Sample_ID = c("SA01"), Cell_count = c(Inf), Age_sample = c(80))
selection_model <- "WGD-chrom-arm-selection"
#------------------------------------------------------CNA PROBABILITIES
prob_CN_whole_genome_duplication <- 0e-4
prob_CN_missegregation <- 5e-5
prob_CN_chrom_arm_missegregation <- 0e-5
alpha_aneuploidy <- 1
formula_CN_missegregation <- "per_homolog:prob_CN_missegregation * ( (alpha_aneuploidy-1)*WGD_count + 1 )"
formula_CN_chrom_arm_missegregation <- "per_homolog:prob_CN_chrom_arm_missegregation * ( (alpha_aneuploidy-1)*WGD_count + 1 )"
#---------------------------------------------------VIABILITY THRESHOLDS
bound_average_ploidy <- Inf
bound_homozygosity <- 0
bound_maximum_CN <- Inf
bound_maximum_CN_normalized <- 2
bound_WGD <- 1
#----------------------------------------------------POPULATION DYNAMICS
vec_time <- T_0[[1]]:T_end[[1]]
L <- 10000
t_0 <- 20
k <- 0.3
vec_cell_count <- L / (1 + exp(-k * (vec_time - t_0)))
table_population_dynamics <- cbind(vec_time, vec_cell_count)
#-----------------------------------------------------------------------\
model_variables_base <- BUILD_general_variables(
    cell_lifespan = cell_lifespan,
    T_0 = T_0, T_end = T_end,
    Table_sample = Table_sample,
    selection_model = selection_model,
    prob_CN_whole_genome_duplication = prob_CN_whole_genome_duplication,
    prob_CN_missegregation = prob_CN_missegregation,
    prob_CN_chrom_arm_missegregation = prob_CN_chrom_arm_missegregation,
    alpha_aneuploidy = alpha_aneuploidy,
    formula_CN_missegregation = formula_CN_missegregation,
    formula_CN_chrom_arm_missegregation = formula_CN_chrom_arm_missegregation,
    bound_maximum_CN = bound_maximum_CN,
    bound_average_ploidy = bound_average_ploidy,
    bound_homozygosity = bound_homozygosity,
    bound_maximum_CN_normalized = bound_maximum_CN_normalized,
    bound_WGD = bound_WGD,
    table_population_dynamics = table_population_dynamics
)
arm_id <- c(paste(model_variables_base$cn_info$Chromosome, "p", sep = ""), paste(model_variables_base$cn_info$Chromosome, "q", sep = ""))
arm_chromosome <- rep(model_variables_base$cn_info$Chromosome, 2)
arm_start <- c(rep(1, length(model_variables_base$cn_info$Chromosome)), model_variables_base$cn_info$Centromere_location + 1)
arm_end <- c(model_variables_base$cn_info$Centromere_location, model_variables_base$cn_info$Bin_count)
arm_s <- rep(1, length(arm_id))

model_variables_base <- BUILD_driver_library(
    model_variables = model_variables_base,
    table_arm_selection_rates = data.frame(Arm_ID = arm_id, Chromosome = arm_chromosome, Bin_start = arm_start, Bin_end = arm_end, s_rate = arm_s)
)

cell_count <- 20
CN_matrix <- BUILD_cn_normal_autosomes(model_variables_base$cn_info)
drivers <- list()
model_variables_base <- BUILD_initial_population(
    model_variables = model_variables_base,
    cell_count = cell_count,
    CN_matrix = CN_matrix,
    drivers = drivers
)
model_variables_base <- CHECK_model_variables(model_variables_base)

model_prefix <- "WGD-and-ANEUPLOIDY"
folder_workplace <- "WGD-and-ANEUPLOIDY"
```

We read the previously fitted parameter values for pan-cancer data.

```{r}
parameter_best <- read.csv("PAN-CANCER_fitted_parameters.csv")
# parameter_best$Best_value[which(parameter_best$Type == "Arm_selection_rate" & parameter_best$Best_value == 1)] <-
#     runif(length(which(parameter_best$Type == "Arm_selection_rate" & parameter_best$Best_value == 1)), 0.9836265, 1.007644)
parameter_best$Best_value[which(parameter_best$Type == "Arm_selection_rate" & parameter_best$Best_value == 1)] <-
    c(0.9920039, 0.9933136, 0.9952504, 0.9859580, 0.9887057, 0.9927065, 0.9871305, 1.0010945, 0.9868608, 0.9964952, 0.9866928, 1.0026826, 0.9981111)



model_variables_base <- bulk_arm_CN_assign_paras(
    model_variables = model_variables_base,
    parameter_IDs = parameter_best$Variable,
    parameters = parameter_best$Best_value
)
tmp <- as.numeric(model_variables_base$general_variables$Value[which(model_variables_base$general_variables$Variable == "prob_CN_missegregation")])
model_variables_base$general_variables$Value[which(model_variables_base$general_variables$Variable == "prob_CN_missegregation")] <-
    1 - (1 - tmp)^(1 / (2 * nrow(model_variables_base$cn_info)))
tmp <- as.numeric(model_variables_base$general_variables$Value[which(model_variables_base$general_variables$Variable == "prob_CN_chrom_arm_missegregation")])
model_variables_base$general_variables$Value[which(model_variables_base$general_variables$Variable == "prob_CN_chrom_arm_missegregation")] <-
    1 - (1 - tmp)^(1 / (2 * nrow(model_variables_base$cn_info)))
```

We set up parameter sweeps of the probability of whole-genome duplication and aneuploidy level.
For every parameter combination, we create 1000 simulations.
We set possible values for the probability of chromosome missegregation to be from 10^{-6}, 2x10^{-6},..., 9x10^{-6}, 10^{-5}.
We set possible values for the probability of aneuploidy to be from 20, 40,..., 180, 200.

```{r}
n_simulations_per_batch <- 1000

var1_name <- "prob_CN_whole_genome_duplication"
var2_name <- "alpha_aneuploidy"

var1_vals <- seq(1e-6, 1e-5, by = 1e-6)
var2_vals <- seq(20, 200, by = 20)
```

We make the simulation run through the possible combinations between the two variables with each single variable combination consisting of 1000 simulations.


```{r}
simulator_multivar(
    model_prefix = model_prefix, folder_workplace = folder_workplace,
    model_variables_base = model_variables_base,
    var1_name = var1_name, var1_vals = var1_vals,
    var2_name = var2_name, var2_vals = var2_vals,
    n_simulations = n_simulations_per_batch,
    stage_final = 3,
    compute_parallel = TRUE
)
```

We obtain the average statistics from the simulations for each parameter value combination.

```{r}
statistics_multivar_matrix(
    model_prefix = model_prefix, folder_workplace = folder_workplace,
    var1_name = var1_name, var1_vals = var1_vals,
    var2_name = var2_name, var2_vals = var2_vals,
    n_simulations = n_simulations_per_batch,
    plot_WGD = TRUE,
    compute_parallel = TRUE
)
```