```{r setup}
# library(readxl)
library(scales)
#-----------------------------------------------------------------------
setwd("/Users/dinhngockhanh/CancerSimulator/R/")
files_sources <- list.files(pattern = "*.r$")
sapply(files_sources, source)
#-----------------------------------------------------------------------
# devtools::install_github("dinhngockhanh/CancerSimulator")
# library(CancerSimulator)
```

First, we set up the fixed parameters for the simulator.
Note that we use the chromosome missegregation per homolog model instead of per division model.

```{r,fig.height=5,fig.width=10}
cell_lifespan <- 1
T_0 <- list(0, "day")
T_end <- list(300, "day")
Table_sample <- data.frame(Sample_ID = c("SA"), Cell_count = c(Inf), Age_sample = c(T_end[[1]]))

T_tau_step <- cell_lifespan / 2
CN_bin_length <- 500000

selection_model <- "chrom-arm-selection"

prob_CN_whole_genome_duplication <- 0
# model_CN_missegregation <- 'per_division'
model_CN_missegregation <- 'per_homolog'
prob_CN_missegregation <- 0
prob_CN_chrom_arm_missegregation <- 0
prob_CN_focal_amplification <- 0
prob_CN_focal_deletion <- 0
prob_CN_cnloh_interstitial <- 0
prob_CN_cnloh_terminal <- 0
model_CN_focal_amplification_length <- "beta"
model_CN_focal_deletion_length <- "beta"
prob_CN_focal_amplification_length_shape_1 <- 0.758304780825031
prob_CN_focal_amplification_length_shape_2 <- 5.33873409782625
prob_CN_focal_deletion_length_shape_1 <- 0.814054548726361
prob_CN_focal_deletion_length_shape_2 <- 6.16614890284825
prob_CN_cnloh_interstitial_length <- 0.005
prob_CN_cnloh_terminal_length <- 0.005
rate_driver <- 0
rate_passenger <- 1e-11

bound_driver <- 3
bound_average_ploidy <- 6
bound_homozygosity <- 0
bound_maximum_CN <- Inf
# bound_maximum_CN <- 8
bound_maximum_CN_normalized <- 4

vec_time <- T_0[[1]]:T_end[[1]]
vec_cell_count <- rep(1000,length(vec_time))
table_population_dynamics <- cbind(vec_time, vec_cell_count)

gc <- read.csv(file = system.file("extdata", "gc_map_500kb.csv", package = "CancerSimulator"))
gc_slope <- 1.2
gc_int <- 0
sigma1 <- 0.02642392
num_reads <- 3906632

model_variables_base <- BUILD_general_variables(
    cell_lifespan = cell_lifespan,
    T_0 = T_0, T_end = T_end, T_tau_step = T_tau_step,
    Table_sample = Table_sample,
    CN_bin_length = CN_bin_length,

    model_CN_missegregation = model_CN_missegregation,

    prob_CN_whole_genome_duplication = prob_CN_whole_genome_duplication,
    prob_CN_missegregation = prob_CN_missegregation,
    prob_CN_chrom_arm_missegregation = prob_CN_chrom_arm_missegregation,
    prob_CN_focal_amplification = prob_CN_focal_amplification,
    prob_CN_focal_deletion = prob_CN_focal_deletion,
    prob_CN_cnloh_interstitial = prob_CN_cnloh_interstitial,
    prob_CN_cnloh_terminal = prob_CN_cnloh_terminal,

    model_CN_focal_amplification_length = model_CN_focal_amplification_length,
    model_CN_focal_deletion_length = model_CN_focal_deletion_length,
    prob_CN_focal_amplification_length_shape_1 = prob_CN_focal_amplification_length_shape_1,
    prob_CN_focal_amplification_length_shape_2 = prob_CN_focal_amplification_length_shape_2,
    prob_CN_focal_deletion_length_shape_1 = prob_CN_focal_deletion_length_shape_1,
    prob_CN_focal_deletion_length_shape_2 = prob_CN_focal_deletion_length_shape_2,
    prob_CN_cnloh_interstitial_length = prob_CN_cnloh_interstitial_length,
    prob_CN_cnloh_terminal_length = prob_CN_cnloh_terminal_length,
    rate_driver = rate_driver,
    rate_passenger = rate_passenger,
    selection_model = selection_model,
    bound_driver = bound_driver,
    bound_average_ploidy = bound_average_ploidy,
    bound_maximum_CN = bound_maximum_CN,
    bound_homozygosity = bound_homozygosity,
    bound_maximum_CN_normalized = bound_maximum_CN_normalized,
    table_population_dynamics = table_population_dynamics,
    gc = gc,
    gc_slope = gc_slope,
    gc_int = gc_int,
    sigma1 = sigma1,
    num_reads = num_reads
)



cancer_type <- "PACA-AU"
fitted_parameters <- read.csv(paste0(cancer_type, "_fitted_parameters.csv"), header = TRUE)
arm_id <- fitted_parameters$Variable[which(fitted_parameters$Type == "Arm_selection_rate")]
arm_chromosome <- rep("", length(arm_id))
arm_start <- rep(0, length(arm_id))
arm_end <- rep(0, length(arm_id))
for (i in 1:length(arm_id)) {
    chrom <- substr(arm_id[i], 1, nchar(arm_id[i]) - 1)
    arm <- substr(arm_id[i], nchar(arm_id[i]), nchar(arm_id[i]))
    if (arm == "p") {
        arm_start[i] <- 1
        arm_end[i] <- model_variables_base$cn_info$Centromere_location[which(model_variables_base$cn_info$Chromosome == chrom)] - 1
    } else if (arm == "q") {
        arm_start[i] <- model_variables_base$cn_info$Centromere_location[which(model_variables_base$cn_info$Chromosome == chrom)]
        arm_end[i] <- model_variables_base$cn_info$Bin_count[which(model_variables_base$cn_info$Chromosome == chrom)]
    }
    arm_chromosome[i] <- chrom
}
arm_s <- fitted_parameters$Best_value[which(fitted_parameters$Type == "Arm_selection_rate")]

model_variables_base <- BUILD_driver_library(
    model_variables = model_variables_base,
    table_arm_selection_rates = data.frame(Arm_ID = arm_id, Chromosome = arm_chromosome, Bin_start = arm_start, Bin_end = arm_end, s_rate = arm_s)
)

cell_count <- 1000
CN_matrix <- BUILD_cn_normal_autosomes(model_variables_base$cn_info)
drivers <- list()
model_variables_base <- BUILD_initial_population(
    model_variables = model_variables_base,
    cell_count = cell_count,
    CN_matrix = CN_matrix,
    drivers = drivers
)
model_variables_base<-CHECK_model_variables(model_variables_base)

model_prefix <- "WGD-vs-MISSEGREGATION"
folder_workplace <- "WGD-vs-MISSEGREGATION"
```


```{r}
# model_variables_base$general_variables$Value[which(model_variables_base$general_variables$Variable == "prob_CN_whole_genome_duplication")] <- 3e-2
# # model_variables_base$general_variables$Value[which(model_variables_base$general_variables$Variable == "prob_CN_missegregation")] <- 1e-2
# model_variables_base$general_variables$Value[which(model_variables_base$general_variables$Variable == "prob_CN_missegregation")] <- 5e-2/46
# model_variables_base<-CHECK_model_variables(model_variables_base)
# tmp <- simulator_full_program(
#     model = model_variables_base,
#     n_simulations = 1,
#     stage_final = 3,
#     save_simulation = FALSE,
#     report_progress = TRUE,
#     seed = 11,
#     compute_parallel = FALSE
# )
```

We set up parameter sweeps of the probability of chromosome missegregation per homolog and probability of whole-genome duplication.
For every parameter combination, we create 5 simulations.
We set possible values for the probability of chromosome missegregation per homolog based on the probability of chromosome miseggregation per division divided by 44, the number of autosomes in a cell.
We keep the probability of whole-genome duplication constant at 3x10^{-2}.

```{r}
n_simulations_per_batch <- 5

var1_name <- "prob_CN_missegregation"
# var1_vals <- seq(1e-3, 1e-2, by = 1e-3)/44
var1_vals <- seq(1e-2,5e-2,by=1e-2)/44
var1_labs <- scientific(var1_vals*44)

var2_name <- "prob_CN_whole_genome_duplication"
var2_vals <- 3e-2
# var2_vals <- 5e-3
var2_labs <- scientific(var2_vals)
# var2_vals <- seq(1e-3, 1e-2, by = 1e-3)
# var2_labs <- scientific(var2_vals)
```

```{r}
simulator_multivar(
    model_prefix = model_prefix, folder_workplace = folder_workplace,
    model_variables_base = model_variables_base,
    var1_name = var1_name, var1_vals = var1_vals, var1_labs = var1_labs,
    var2_name = var2_name, var2_vals = var2_vals, var2_labs = var2_labs,
    n_simulations = n_simulations_per_batch,
    stage_final = 3,
    compute_parallel = TRUE
)
```

```{r}
# statistics_multivar_matrix(
#     model_prefix = model_prefix, folder_workplace = folder_workplace,
#     var1_name = var1_name, var1_vals = var1_vals, var1_labs = var1_labs,
#     var2_name = var2_name, var2_vals = var2_vals, var2_labs = var2_labs,
#     n_simulations = n_simulations_per_batch,
#     plot_WGD = TRUE,
#     plot_misseg = TRUE,
#     compute_parallel = TRUE
# )
```

We can see how the statistics change by keeping the probability of whole-genome duplication constant for all values of the probability of chromosomal missegregation per homolog.

```{r}
var2_fixed <- var2_labs[1]
statistics_multivar_vector(
    model_prefix = model_prefix, folder_workplace = folder_workplace,
    var1_name = var1_name, var1_labs = var1_labs,
    var2_name = var2_name, var2_labs = rep(var2_fixed, length(var1_labs)),
    var_labs = var1_labs, name_lab = "Probability of missegregation",
    plot_WGD = TRUE,
    n_simulations = n_simulations_per_batch,
    plotname = paste0(model_prefix, "_prob_CN_missegregation_varies_and_prob_CN_whole_genome_duplication=", var2_fixed),
    example_simulation = FALSE
)
```

```{r}
# var1_fixed <- var1_labs[length(var1_labs)]
# statistics_multivar_vector(
#     model_prefix = model_prefix, folder_workplace = folder_workplace,
#     var1_name = var1_name, var1_labs = rep(var1_fixed, length(var2_labs)),
#     var2_name = var2_name, var2_labs = var2_labs,
#     var_labs = var2_labs, name_lab = "Probability of WGD",
#     plot_WGD = TRUE,
#     n_simulations = n_simulations_per_batch,
#     plotname = paste0(model_prefix, "_prob_CN_whole_genome_duplication_varies_and_prob_CN_missegregation=", var1_fixed),
#     example_simulation = FALSE
# )
```



```{r}
```
