```{r setup}
devtools::install_github("dinhngockhanh/CancerSimulator")


library(ggplot2)
library(signals)
library(CancerSimulator)

devtools::load_all()
devtools::document()
```

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% MIXTURE EXPERIMENT
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

```{r}
# for(i in 1:100){
# print(paste('++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++   SIMULATION ',i,sep=''))

print('================================================================================   PART A')
MODEL                         <- 'MIXTURE-A'
stage_final                   <- 3
package_output_A              <- SIMULATOR_FULL_PROGRAM_one_simulation(MODEL,stage_final)

package_clonal_evolution_A    <- package_output_A[[1]]
package_sample_A              <- package_output_A[[2]]
package_sample_phylogeny_A    <- package_output_A[[3]]

sample_genotype_profiles_A    <- package_sample_A[[1]]
phylogeny_clustering_truth_A  <- package_sample_phylogeny_A[[1]]

plotHeatmap(sample_genotype_profiles_A,plotcol="state",clusters=phylogeny_clustering_truth_A$clustering,tree=phylogeny_clustering_truth_A$tree,reorderclusters=TRUE,plottree=TRUE)

plotHeatmap(sample_genotype_profiles_A,plotcol="Min",clusters=phylogeny_clustering_truth_A$clustering,tree=phylogeny_clustering_truth_A$tree,reorderclusters=TRUE,plottree=TRUE)
print('================================================================================   PREPARE INITIAL STATE FOR EXPERIMENTS B AND C')
MODEL_NEW                     <- 'MIXTURE-B'
EXTRA_build_model_variables_from_sample(MODEL_NEW,package_output_A)
MODEL_NEW                     <- 'MIXTURE-C'
EXTRA_build_model_variables_from_sample(MODEL_NEW,package_output_A)
print('================================================================================   PART B')
MODEL                         <- 'MIXTURE-B'
stage_final                   <- 3
package_output_B              <- SIMULATOR_FULL_PROGRAM_one_simulation(MODEL,stage_final)

package_clonal_evolution_B    <- package_output_B[[1]]
package_sample_B              <- package_output_B[[2]]
package_sample_phylogeny_B    <- package_output_B[[3]]

sample_genotype_profiles_B    <- package_sample_B[[1]]
phylogeny_clustering_truth_B  <- package_sample_phylogeny_B[[1]]

# plotHeatmap(sample_genotype_profiles_B,plotcol="state",clusters=phylogeny_clustering_truth_B$clustering,tree=phylogeny_clustering_truth_B$tree,reorderclusters=TRUE,plottree=TRUE)
#
# plotHeatmap(sample_genotype_profiles_B,plotcol="Min",clusters=phylogeny_clustering_truth_B$clustering,tree=phylogeny_clustering_truth_B$tree,reorderclusters=TRUE,plottree=TRUE)
print('================================================================================   PART C')
MODEL                         <- 'MIXTURE-C'
stage_final                   <- 3
package_output_C              <- SIMULATOR_FULL_PROGRAM_one_simulation(MODEL,stage_final)


package_clonal_evolution_C    <- package_output_C[[1]]
package_sample_C              <- package_output_C[[2]]
package_sample_phylogeny_C    <- package_output_C[[3]]

sample_genotype_profiles_C    <- package_sample_C[[1]]
phylogeny_clustering_truth_C  <- package_sample_phylogeny_C[[1]]

# plotHeatmap(sample_genotype_profiles_C,plotcol="state",clusters=phylogeny_clustering_truth_C$clustering,tree=phylogeny_clustering_truth_C$tree,reorderclusters=TRUE,plottree=TRUE)
#
# plotHeatmap(sample_genotype_profiles_C,plotcol="Min",clusters=phylogeny_clustering_truth_C$clustering,tree=phylogeny_clustering_truth_C$tree,reorderclusters=TRUE,plottree=TRUE)
# }
```



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% NEUTRAL FALLOPIAN TUBES - ONE SIMULATION
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

```{r}
MODEL                         <- 'FALLOPIAN-TUBES-NEUTRAL'
stage_final                   <- 3



package_simulation            <- SIMULATOR_FULL_PROGRAM_one_simulation(MODEL,stage_final)
package_sample                <- package_simulation[[2]]
sample_genotype_profiles      <- package_sample[[1]]



head(sample_genotype_profiles)



# vec_time_plot                 <- seq(from=0,to=80*365,by=365)
PLOT_total_population_vs_input(package_simulation,MODEL,vec_time_plot=seq(0,80,by=1),unit='year')




# package_sample_phylogeny      <- package_simulation[[3]]
# phylogeny_clustering_truth    <- package_sample_phylogeny[[1]]
#
# plotHeatmap(sample_genotype_profiles,plotcol="state",clusters=phylogeny_clustering_truth$clustering,tree=phylogeny_clustering_truth$tree,reorderclusters=TRUE,plottree=TRUE)
#
# plotHeatmap(sample_genotype_profiles,plotcol="Min",clusters=phylogeny_clustering_truth$clustering,tree=phylogeny_clustering_truth$tree,reorderclusters=TRUE,plottree=TRUE)
```

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%% NEUTRAL FALLOPIAN TUBES - MANY SIMULATIONS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

```{r}
MODEL                         <- 'FALLOPIAN-TUBES-NEUTRAL'
stage_final                   <- 1
N_simulations                 <- 100
SIMULATOR_FULL_PROGRAM_many_simulations(MODEL,stage_final,N_simulations)
```

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% SELECTION OF BULK HGSOC - ONE SIMULATION
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

```{r}
MODEL <- 'HGSOC-BULK'
SIMULATOR_VARIABLES_for_simulation(MODEL)

parameter_set                                         <- list()
parameter_set$vec_selection_rates                     <- rep(0.05,nrow(driver_library))
parameter_set$rate_driver                             <- 1e-16
parameter_set$prob_CN_whole_genome_duplication        <- 1e-6
parameter_set$prob_CN_missegregation                  <- 1e-6
parameter_set$prob_CN_chrom_arm_missegregation        <- 1e-6
parameter_set$prob_CN_focal_amplification             <- 1e-6
parameter_set$prob_CN_focal_deletion                  <- 1e-6
parameter_set$prob_CN_cnloh_interstitial              <- 1e-6
parameter_set$prob_CN_cnloh_terminal                  <- 1e-6
parameter_set$prob_CN_focal_amplification_length      <- 0.1
parameter_set$prob_CN_focal_deletion_length           <- 0.1
parameter_set$prob_CN_cnloh_interstitial_length       <- 0.1
parameter_set$prob_CN_cnloh_terminal_length           <- 0.1

ABC_one_simulation(MODEL,parameter_set)
```

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%% SELECTION OF BULK HGSOC - MANY SIMULATIONS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

```{r}
MODEL <- 'HGSOC-BULK'
SIMULATOR_VARIABLES_for_simulation(MODEL)

parameter_set                                         <- list()
parameter_set$vec_selection_rates                     <- rep(0.05,nrow(driver_library))
parameter_set$rate_driver                             <- 1e-16
parameter_set$prob_CN_whole_genome_duplication        <- 1e-6
parameter_set$prob_CN_missegregation                  <- 1e-6
parameter_set$prob_CN_chrom_arm_missegregation        <- 1e-6
parameter_set$prob_CN_focal_amplification             <- 1e-6
parameter_set$prob_CN_focal_deletion                  <- 1e-6
parameter_set$prob_CN_cnloh_interstitial              <- 1e-6
parameter_set$prob_CN_cnloh_terminal                  <- 1e-6
parameter_set$prob_CN_focal_amplification_length      <- 0.1
parameter_set$prob_CN_focal_deletion_length           <- 0.1
parameter_set$prob_CN_cnloh_interstitial_length       <- 0.1
parameter_set$prob_CN_cnloh_terminal_length           <- 0.1

N_simulations                                         <- 100

ABC_many_simulations(MODEL,parameter_set,N_simulations)
```
