---
title: "How to use Cancer Simulator"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{How to use Cancer Simulator}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  warning = FALSE, message = FALSE,
  comment = "#>"
)
```

```{r setup}
library(ggplot2)
library(ggtree)
library(signals)
library(dendextend)
library(fishplot)
library(ctc)
library(adephylo)
library(data.table)
library(HMMcopy)
library(plyr)
library(matrixStats)
library(parallel)
library(RColorBrewer)
library(ape)
#-----------------------------------------------------------------------
# setwd("/Users/dinhngockhanh/CancerSimulator/R/")
# files_sources <- list.files(pattern = "*.r$")
# sapply(files_sources, source)
#-----------------------------------------------------------------------
devtools::install_github("dinhngockhanh/CancerSimulator")
library(CancerSimulator)
```

# Set up model variables

Cancer Simulator requires several csv files containing variables to set up the simulations. Here we detail how to easily build these files.

## General variables for every cell

The first step is to specify the variables describing every cell during the process. The cells are assumed to have the same expected lifespan (regardless of their driver or CN profiles). The cell lifespan is always in days.

```{r}
cell_lifespan                       <- 30
```

Next, we specify the times at the beginning and the end of the simulation. The unit can be `day`, `week`, `month` or `year`.

```{r}
T_0                                 <- list(0,'year')
T_end                               <- list(80,'year')
```

We

We will sample 200 cells from the population at `T_end`:

```{r}
# Table_sample <- data.frame(Sample_ID=c('SA01','SA02','SA03'),Cell_count=c(100,100,100),Age_sample=c(40,60,80))
Table_sample                        <- data.frame(Sample_ID='SA01',Cell_count=300,Age_sample=80)
```

We use the tau-leaping algorithm to simulate the clonal evolution. This requires a time step. The smaller this step is, the closer the resulting simulation is to the traditional Gillespie's algorithm, but it also means increasing runtime. By default, we use tau = 3 days.

```{r}
T_tau_step                          <- cell_lifespan/2
```

An important variable is the size of CN bins. By default, this will be 500,000 bp.

```{r}
CN_bin_length                       <- 500000
```

We have to set up the probability that each class of CN events occurs in a cell division. The local events (chromosome, chromosome-arm or focal level), if happening, will occur on a random location among eligible places.

```{r}
prob_CN_whole_genome_duplication    <- 0.5e-4
prob_CN_missegregation              <- 2e-4
prob_CN_chrom_arm_missegregation    <- 2e-4
prob_CN_focal_amplification         <- 0.e-4
prob_CN_focal_deletion              <- 2e-4
prob_CN_cnloh_interstitial          <- 0e-4
prob_CN_cnloh_terminal              <- 0e-4
```

The focal and CNLOH events require the geometric parameters for the length (in bin count) of the DNA region being affected.

```{r}
prob_CN_focal_amplification_length  <- 0.5
prob_CN_focal_deletion_length       <- 0.5
prob_CN_cnloh_interstitial_length   <- 0.5
prob_CN_cnloh_terminal_length       <- 0.5
```

Next, we set up the mutation rates for drivers and passengers, per bp per cell division.

```{r}
rate_driver                         <- 0
# rate_driver                         <- 8e-16
rate_passenger                      <- 1e-11
```

We can also set up the upper limits for the driver count and local CN for a cell to still be viable.

```{r}
selection_model                     <- 'old'
bound_driver                        <- 3
bound_average_ploidy                <- 4.5
bound_homozygosity                  <- 500
```

Finally, we need to set up the growth model for the total cell population. The algorithm is designed to follow any dynamics (within the limit imposed by how fast the cells can divide, defined by their lifespan). Note that to avoid cases where the simulations are terminated prematurely because of the stochasticity, the population size should not decrease to below about 10 cells at any time.

We build one vector for the time points (in the same unit as `T_0` and `T_end`, therefore in years in this case), and another vector for the corresponding expected population sizes. As an example, we use the logistic function.

```{r}
vec_time                            <- T_0[[1]]:T_end[[1]]
L                                   <- 10000
t_0                                 <- 20
k                                   <- 0.3
vec_cell_count                      <- L/(1+exp(-k*(vec_time-t_0)))
table_population_dynamics           <- cbind(vec_time,vec_cell_count)
```

Parameters for the sequencing noise and bias model:

```{r}
gc                                  <- read.csv(file = system.file("extdata", "gc_map_500kb.csv", package = "CancerSimulator"))
gc_slope                            <- 1.2
gc_int                              <- 0
sigma1                              <- 0.1
num_reads                           <- 2e6
```

Finally, we build the model variable files, using function `BUILD_general_variables`:

```{r}
model_variables   <- BUILD_general_variables(
    cell_lifespan=cell_lifespan,
    T_0=T_0,T_end=T_end,T_tau_step=T_tau_step,
    Table_sample=Table_sample,
    CN_bin_length=CN_bin_length,
    prob_CN_whole_genome_duplication=prob_CN_whole_genome_duplication,
    prob_CN_missegregation=prob_CN_missegregation,
    prob_CN_chrom_arm_missegregation=prob_CN_chrom_arm_missegregation,
    prob_CN_focal_amplification=prob_CN_focal_amplification,
    prob_CN_focal_deletion=prob_CN_focal_deletion,
    prob_CN_cnloh_interstitial=prob_CN_cnloh_interstitial,
    prob_CN_cnloh_terminal=prob_CN_cnloh_terminal,
    prob_CN_focal_amplification_length=prob_CN_focal_amplification_length,
    prob_CN_focal_deletion_length=prob_CN_focal_deletion_length,
    prob_CN_cnloh_interstitial_length=prob_CN_cnloh_interstitial_length,
    prob_CN_cnloh_terminal_length=prob_CN_cnloh_terminal_length,
    rate_driver=rate_driver,
    rate_passenger=rate_passenger,
    selection_model=selection_model,
    bound_driver=bound_driver,
    bound_average_ploidy=bound_average_ploidy,
    bound_homozygosity=bound_homozygosity,
    table_population_dynamics=table_population_dynamics,
    gc=gc,
    gc_slope=gc_slope,
    gc_int=gc_int,
    sigma1=sigma1,
    num_reads=num_reads)
```

## Driver library

We first define a list of drivers to include in the model:

```{r}
vec_driver_genes  <- c('TP53','BRCA1','BRCA2','MSH2','MSH6','ARID1A',
'STK11','BRAF','MLH1','PIK3R1','ERBB2','CTNNB1','AKT1','PPP2R1A')
```

For each driver, we need to decide if it behaves as a TSG or an oncogene. Note that the algorithm requires at least one gene of each type to behave properly.

```{r}
vec_driver_role   <- c('TSG','TSG','TSG','TSG','TSG','TSG',
'TSG','ONCOGENE','TSG','TSG','ONCOGENE','ONCOGENE','ONCOGENE','TSG')
```

We then give each driver a selective strength. The higher it is, the more likely the mutated cell will escape death and the faster it reaches fixation.

```{r}
vec_driver_s      <- rep(0.003,length(vec_driver_genes))
vec_driver_s[1]   <- 0.01
vec_driver_s[2]   <- 0.005
vec_driver_s[3]   <- 0.005
```

The function `BUILD_driver_library` builds the driver library. If the chromosome and bin locations of the drivers are not given, the function will find them by crosschecking with the Cancer Gene Census database.

```{r}
model_variables   <- BUILD_driver_library(model_variables=model_variables,vec_driver_genes=vec_driver_genes,vec_driver_role=vec_driver_role,vec_driver_s=vec_driver_s)
```

## Initial cell population

Assume we have 20 cells at `T_0`, which are all normal diploid cells. We will declare the specifics of each clone.

Let's start by declaring the clone for the first 10 cells. `cell_count` tells the algorithm how many cells there are in this clone at `T_0`.

```{r}
cell_count        <- 20
```

`CN_matrix` is a data frame with five variables, in which `Allele` gives the allele specification for the region from `Bin_start` to `Bin_end` of `Strand` of `Chromosome`. The user can build this data frame themselves, or use function `BUILD_cn_normal_XX` to build the normal diploid CN profile.

```{r}
CN_matrix         <- BUILD_cn_normal_XX(model_variables$cn_info)
```

Finally, `drivers` lists the mutated drivers. If there is no driver, it is an empty list.

```{r}
drivers           <- list()
```

We use the function `BUILD_initial_population` to add this clone to the initial cell population.

```{r}
model_variables   <- BUILD_initial_population(model_variables=model_variables,cell_count=cell_count,CN_matrix=CN_matrix,drivers=drivers)
```

## Save the input files

The last step is to save the model variable files.

Let's give a name to the model.

```{r}
model_name        <- 'EXPERIMENT-1A-SELECTIVE'
```

The function `SAVE_model_variables` saves the input variables required to run the simulations according to our specification in 7 csv files. The names for these files begin with `MODEL_NAME`.

```{r}
SAVE_model_variables(model_name=model_name,model_variables=model_variables)
```

# Run the simulation

The function `SIMULATOR_FULL_PROGRAM_one_simulation` produces one simulation. It will look for the csv files from above to set the model variables.

`stage_final` dictates how far the simulation will go:
* `stage_final`=1 only produces the clonal evolution.
* `stage_final`=2 samples cells as specified above, and outputs their CN profiles as a `rda` file.
* `stage_final`=3 further builds the phylogeny for these cells, and outputs this as a `rda` file. The phylogeny is built based on information from the clonal evolution.

`N_clones_min` and `N_clones_max` are optional variables in case we want to limit the number of clones in the simulation.

```{r}
n_simulations     <- 1
stage_final       <- 3
n_clones_min      <- 1
n_clones_max      <- Inf

tmp               <- simulator_full_program(
    model = model_name,
    n_simulations = n_simulations,
    stage_final = stage_final,
    n_clones_min = n_clones_min,
    n_clones_max = n_clones_max,
    save_simulation = TRUE,
    internal_nodes_cn_info = TRUE,
    save_newick_tree = TRUE,

    save_cn_profile = TRUE,
    save_cn_clones = TRUE,
    model_readcount = TRUE,
    format_cn_profile = "both",

    apply_HMM = FALSE,
    apply_UMAP_on_HMM = FALSE,

    report_progress = TRUE,
    seed = 10
)
```

A variety of plots from the simulation:

```{r}
plot_all(model = model_name,n_simulations = n_simulations)
```
